// @unbndl/view/src/view.ts
import {
  Template,
  TemplateArgs,
  TemplateValue,
  createScope
} from "@unbndl/html";

export function createView<T extends object>(
  html: Template<[T]>
) {
  return html;
}

export function createView2<
  U extends object,
  V extends object
>(html: Template<[U, V]>) {
  return html;
}

export function createViewN<TT extends TemplateArgs>(
  html: Template<TT>
) {
  return html;
}

type ArrayArgs<TT extends TemplateArgs> = {
  [Index in keyof TT]: Array<TT[Index]>;
};

function zipArgs<TT extends TemplateArgs>(
  ...lists: ArrayArgs<TT>
): Array<TT> {
  const maxLength = lists
    .map((arr) => arr.length)
    .reduce((a, b) => Math.max(a, b), 0);
  const range = Array.from(Array(maxLength).keys());
  return range.map((i) => lists.map((arr) => arr[i]) as TT);
}

function map<T extends object>(
  view: Template<[T]>,
  list: Array<T>
): TemplateValue<[T]> {
  return mapN(view, list);
}

function map2<
  U extends object,
  V extends object = {}
>(
  view: Template<[U, V]>,
  ulist: Array<U>,
  vlist: Array<V>
): TemplateValue<[U, V]> {
  return mapN(view, ulist, vlist);
}

function mapN<TT extends TemplateArgs>(
  view: Template<TT>,
  ...lists: ArrayArgs<TT>
): TemplateValue<TT> {
  if (!lists.length) return "";

  return zipArgs<TT>(...lists)
    .map((tuple: TT) => applyN(view, ...tuple))
    .flat();
}

function apply<T extends object>(
  view: Template<[T]>,
  t: T | undefined
): TemplateValue<[T]> {
  return typeof t === "undefined" ? "" : applyN(view, t);
}

function apply2<
  U extends object,
  V extends object = {}
>(
  view: Template<[U, V]>,
  u: U | undefined,
  v: V | undefined
): TemplateValue<[U, V]> {
  return typeof u === "undefined" || typeof v === "undefined"
    ? ""
    : applyN(view, u, v);
}

function applyN<TT extends TemplateArgs>(
  view: Template<TT>,
  ...tuple: TT
): TemplateValue<TT> {
  // console.log("Applying viewmodel", $);
  if (!tuple) return "";
  const scope = createScope(tuple);
  return view.render(...scope);
}

export const View = {
  apply,
  apply2,
  applyN,
  map,
  map2,
  mapN
};
