// @unbndl/view/src/viewModel.ts
import { MappedSource, NameMapping, Source } from "./source.ts";
import { Context, Template  } from "@unbndl/html";

export type ViewState<T = object> = {
  [K in keyof Partial<T>]: any;
};

export class ViewModel<
  T extends ViewState<T>
> extends Context<T> {
  constructor(init: T, adoptedContext?: Context<T>) {
    super(init, adoptedContext);
  }

  get $(): Readonly<T> {
    return this.toObject();
  }

  with<S extends ViewState<T> = T>(
    source: Source<S>,
    ...keys: Array<keyof S & keyof T>
  ): ViewModel<T> {
    const mapping: NameMapping<S, S> = Object.fromEntries(
      keys.map((k: keyof S) => [k, k]) as Array<Array<keyof S>>
    );
    return this.merge(new MappedSource<S, T>(source, mapping));
  }

  withCalculated<S extends ViewState>(
    source: Source<S>,
    mapping: NameMapping<T, S>
  ): ViewModel<T> {
    return this.merge(new MappedSource<S, T>(source, mapping));
  }

  withRenamed<S extends ViewState>(
    source: Source<S>,
    renaming: { [K in keyof Partial<T>]: keyof S }
  ): ViewModel<T> {
    return this.merge(new MappedSource<S, T>(source, renaming));
  }

  merge<S extends ViewState<T>>(
    source: Source<S>
  ): ViewModel<T> {
    if (source) {
      const initialized: Map<keyof S, boolean> = new Map();
      source
        .start((name: keyof S, value: any) => {
          this.set(name, value);
          initialized.set(name, true);
        })
        .then((firstObservation: Partial<S>) => {
          const keys = (Object.keys(firstObservation) as Array<
            keyof S>)
            .filter((k) => !initialized.get(k));

          keys.forEach((k: keyof S) => {
            // console.log("First observation of source:", firstObservation)
            this.set(k, firstObservation[k] as S[typeof k])
          }
          );
        });
    }

    return this;
  }

  render(template: Template<[T]>): DocumentFragment {
    // console.log("📷 Rendering view, context=", this);
    return template.render(this as Context<T>);
  }
}

export function createViewModel<
  T extends object
>(): {} extends T ? ViewModel<T> : never;
export function createViewModel<T extends object>(
  init: T
): ViewModel<T>;

export function createViewModel<T extends object>(init?: T) {
  if (init === undefined)
    return new ViewModel<T>({} as Required<T>);
  else return new ViewModel<T>(init);
}
