// @unbndl/html/src/effects/manager.ts
import { Effect } from "./effect";
import { SignalEvent } from "./signal";
import { Scheduler } from "./scheduler";

export class EffectsManager<T extends object> {
  private running: Effect[] = [];
  private host?: EventTarget;
  private eventType = "un-effect:change";

  isRunning(): boolean {
    return this.running.length > 0;
  }

  push(effect: Effect): void {
    // console.log("Starting manager for effect", effect);
    this.running.push(effect);
  }

  pop(): void {
    // console.log("Stopping manager for effect");
    this.running.pop();
  }

  current(): Effect | undefined {
    const len = this.running.length;
    return len ? this.running[len - 1] : undefined;
  }

  subscribe(key: keyof T, scope: T): void {
    const current = this.current();
    // console.log("Subscribing to signal", key, scope, !!current);
    if (current) {
      Scheduler.scheduler.subscribe(scope, key, current);
    }
  }

  runEffects(key: keyof T, scope: T): void {
    console.log("⚙️ Running effects for signal", key, scope);
    Scheduler.scheduler.scheduleEffects(scope, key);
    if (this.host) {
      const evt = new SignalEvent(this.eventType, {
        property: key,
        value: scope[key]
      });
      this.host.dispatchEvent(evt);
    }
  }

  setHost(host: EventTarget, eventType?: string) {
    this.host = host;
    if (eventType) this.eventType = eventType;
  }
}
