// @unbndl/html/src/effects/scheduler.ts
import { Scope } from "./scope";
import { Effect, Effector, EffectArgs } from "./effect";

export function createEffect<TT extends EffectArgs>(
  fn: Effector<TT>,
  ...scope: Scope<TT>
): void {
  const effect = {
    execute() {
      console.log("▶️ Executing created effect:", scope, fn);
      const args = scope.map((cx) => cx.open(effect)) as TT;
      fn(...args);
      scope.forEach((cx) => cx.close());
    }
  };

  effect.execute();
}

export type Signals<T extends object> = Map<
  keyof T,
  Set<Effect>
>;

export class Scheduler {
  private signals: WeakMap<object, Signals<object>> =
    new WeakMap();
  private scheduled: WeakSet<Effect> = new WeakSet();

  static scheduler = new Scheduler();

  subscribe<T extends object>(
    scope: T,
    key: keyof T,
    effect: Effect
  ) {
    let signals: Signals<T> | undefined =
      this.signals.get(scope);
    if (!signals) {
      signals = new Map() as Signals<T>;
      this.signals.set(scope, signals as Signals<object>);
    }
    let signal = signals.get(key);
    if (!signal) {
      signal = new Set();
      signals.set(key, signal);
    }
    signal.add(effect);
  }

  scheduleEffects<T extends object>(scope: T, key: keyof T) {
    const signals: Signals<T> | undefined =
      this.signals.get(scope);

    if (!signals) return;
    const signal = signals.get(key);
    if (signal) {
      for (const effect of signal) {
        if (!this.scheduled.has(effect)) {
          this.scheduled.add(effect);
          setTimeout(() => {
            this.scheduled.delete(effect);
            effect.execute();
          });
        }
      }
    }
  }
}
