// @unbndl/switch/dist/switch.umd.cjs
(function(global, factory) {
	typeof exports === "object" && typeof module !== "undefined" ? factory(exports) : typeof define === "function" && define.amd ? define(["exports"], factory) : (global = typeof globalThis !== "undefined" ? globalThis : global || self, factory(global.switch = {}));
})(this, function(exports) {
	Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
	//#region \0rolldown/runtime.js
	var __create = Object.create;
	var __defProp$3 = Object.defineProperty;
	var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
	var __getOwnPropNames = Object.getOwnPropertyNames;
	var __getProtoOf = Object.getPrototypeOf;
	var __hasOwnProp = Object.prototype.hasOwnProperty;
	var __commonJSMin = (cb, mod) => () => (mod || (cb((mod = { exports: {} }).exports, mod), cb = null), mod.exports);
	var __exportAll$3 = (all, no_symbols) => {
		let target = {};
		for (var name in all) __defProp$3(target, name, {
			get: all[name],
			enumerable: true
		});
		if (!no_symbols) __defProp$3(target, Symbol.toStringTag, { value: "Module" });
		return target;
	};
	var __copyProps = (to, from, except, desc) => {
		if (from && typeof from === "object" || typeof from === "function") for (var keys = __getOwnPropNames(from), i = 0, n = keys.length, key; i < n; i++) {
			key = keys[i];
			if (!__hasOwnProp.call(to, key) && key !== except) __defProp$3(to, key, {
				get: ((k) => from[k]).bind(null, key),
				enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
			});
		}
		return to;
	};
	var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(isNodeMode || !mod || !mod.__esModule ? __defProp$3(target, "default", {
		value: mod,
		enumerable: true
	}) : target, mod));
	//#endregion
	//#region ../service/dist/service.js
	var __defProp$2 = Object.defineProperty;
	var __exportAll$2 = (all, no_symbols) => {
		let target = {};
		for (var name in all) __defProp$2(target, name, {
			get: all[name],
			enumerable: true
		});
		if (!no_symbols) __defProp$2(target, Symbol.toStringTag, { value: "Module" });
		return target;
	};
	var message_exports$1 = /* @__PURE__ */ __exportAll$2({
		Dispatch: () => Dispatch$1,
		None: () => None$1,
		dispatch: () => dispatch$3,
		dispatcher: () => dispatcher$1
	});
	var None$1 = [];
	var Dispatch$1 = class extends CustomEvent {
		constructor(msg, eventType = "un:message") {
			super(eventType, {
				bubbles: true,
				composed: true,
				detail: msg
			});
		}
	};
	function dispatcher$1(eventType = "un:message") {
		return (target, ...msg) => target.dispatchEvent(new Dispatch$1(msg, eventType));
	}
	var dispatch$3 = dispatcher$1();
	var SignalEvent$3 = class extends CustomEvent {
		constructor(eventType, signal) {
			super(eventType, {
				bubbles: true,
				cancelable: true,
				composed: true,
				detail: signal
			});
		}
	};
	function createEffect$3(fn, ...scope) {
		const effect = { execute() {
			console.log("▶️ Executing created effect:", scope, fn);
			fn(...scope.map((cx) => cx.open(effect)));
			scope.forEach((cx) => cx.close());
		} };
		effect.execute();
	}
	var Scheduler$3 = class Scheduler {
		constructor() {
			this.signals = /* @__PURE__ */ new WeakMap();
			this.scheduled = /* @__PURE__ */ new WeakSet();
		}
		static {
			this.scheduler = new Scheduler();
		}
		subscribe(scope, key, effect) {
			let signals = this.signals.get(scope);
			if (!signals) {
				signals = /* @__PURE__ */ new Map();
				this.signals.set(scope, signals);
			}
			let signal = signals.get(key);
			if (!signal) {
				signal = /* @__PURE__ */ new Set();
				signals.set(key, signal);
			}
			signal.add(effect);
		}
		scheduleEffects(scope, key) {
			const signals = this.signals.get(scope);
			if (!signals) return;
			const signal = signals.get(key);
			if (signal) {
				for (const effect of signal) if (!this.scheduled.has(effect)) {
					this.scheduled.add(effect);
					setTimeout(() => {
						this.scheduled.delete(effect);
						effect.execute();
					});
				}
			}
		}
	};
	var EffectsManager$3 = class {
		constructor() {
			this.running = [];
			this.eventType = "un-effect:change";
		}
		isRunning() {
			return this.running.length > 0;
		}
		push(effect) {
			this.running.push(effect);
		}
		pop() {
			this.running.pop();
		}
		current() {
			const len = this.running.length;
			return len ? this.running[len - 1] : void 0;
		}
		subscribe(key, scope) {
			const current = this.current();
			if (current) Scheduler$3.scheduler.subscribe(scope, key, current);
		}
		runEffects(key, scope) {
			console.log("⚙️ Running effects for signal", key, scope);
			Scheduler$3.scheduler.scheduleEffects(scope, key);
			if (this.host) {
				const evt = new SignalEvent$3(this.eventType, {
					property: key,
					value: scope[key]
				});
				this.host.dispatchEvent(evt);
			}
		}
		setHost(host, eventType) {
			this.host = host;
			if (eventType) this.eventType = eventType;
		}
	};
	var Context$3 = class {
		static {
			this.CHANGE_EVENT_TYPE = "un-context:change";
		}
		constructor(init, adoptedContext) {
			if (adoptedContext) {
				this.manager = adoptedContext.manager;
				this.object = adoptedContext.object;
				this.proxy = adoptedContext.proxy;
				this.update(init);
			} else {
				this.manager = new EffectsManager$3();
				this.object = init;
				this.proxy = createContext$3(this.object, this.manager);
			}
		}
		get(prop) {
			return this.proxy[prop];
		}
		set(prop, value) {
			this.proxy[prop] = value;
		}
		toObject() {
			return this.proxy;
		}
		update(next) {
			Object.assign(this.proxy, next);
		}
		apply(mapFn) {
			this.update(mapFn(this.proxy));
		}
		createEffect(fn) {
			createEffect$3(fn, this);
		}
		setHost(host, eventType) {
			this.manager.setHost(host, eventType);
		}
		open(effect) {
			this.manager.push(effect);
			return this.proxy;
		}
		close() {
			this.manager.pop();
		}
	};
	function createContext$3(root, manager) {
		return new Proxy(root, {
			get: (subject, prop, receiver) => {
				const value = Reflect.get(subject, prop, receiver);
				if (manager.isRunning() && isObservable$3(value)) manager.subscribe(prop, subject);
				return value;
			},
			set: (subject, prop, newValue, receiver) => {
				const didSet = Reflect.set(subject, prop, newValue, receiver);
				if (didSet && isObservable$3(newValue)) manager.runEffects(prop, subject);
				return didSet;
			}
		});
	}
	function isObservable$3(value) {
		switch (typeof value) {
			case "object":
			case "number":
			case "string":
			case "symbol":
			case "boolean":
			case "undefined": return true;
			default: return false;
		}
	}
	var DirectEffect$1 = class {
		constructor(fn, ...scope) {
			this.effectFn = () => fn(...scope);
		}
		execute() {
			this.effectFn();
		}
	};
	function createTemplate$3(fragment, render) {
		Object.assign(fragment, { render });
		return fragment;
	}
	function renderForEffects$3(original, effectors, ...scope) {
		const fragment = original.cloneNode(true);
		Array.from(effectors.entries()).forEach(([label, mutations]) => {
			const site = fragment.querySelector(`[data-${label}]`);
			if (site) mutations.forEach((fn) => fn(site, fragment, ...scope));
		});
		return fragment;
	}
	var TemplateParser$3 = class TemplateParser {
		static {
			this.parser = new DOMParser();
		}
		constructor(docType) {
			this.docType = "text/html";
			this.plugins = [];
			if (docType) this.docType = docType;
		}
		use(plugin) {
			this.plugins = this.plugins.concat(plugin);
		}
		parse(template, params) {
			const postProcess = {};
			const stringToParse = template.map((s, i) => {
				if (i >= params.length) return [s];
				const param = params[i];
				const place = this.classifyPlace(i, template);
				const mutation = this.tryReplacements(place, param);
				if (mutation) {
					const post = postProcess[place.nodeLabel];
					if (post) post.push(mutation);
					else postProcess[place.nodeLabel] = [mutation];
					switch (place.kind) {
						case "attr value": return [s, `"" data-${place.nodeLabel}`];
						case "tag content": return [s, `data-${place.nodeLabel}`];
						case "element content": return [s, `<ins data-${place.nodeLabel}></ins>`];
					}
				} else {
					console.error("No match for template parameter", place, param);
					throw `Failed to render template parameter ${i} around ${s}`;
				}
				return [s];
			}).flat().join("");
			const doc = TemplateParser.parser.parseFromString(stringToParse, this.docType);
			const collection = doc.head.childElementCount ? doc.head.children : doc.body.children;
			const fragment = new DocumentFragment();
			fragment.replaceChildren(...collection);
			const effectors = /* @__PURE__ */ new Map();
			for (const label in postProcess) {
				const site = fragment.querySelector(`[data-${label}]`);
				if (site) postProcess[label].forEach((m) => {
					const effector = m.apply(site, fragment);
					if (effector) {
						let list = effectors.get(label);
						if (list) list.push(effector);
						else effectors.set(label, [effector]);
					}
				});
			}
			return createTemplate$3(fragment, (...scope) => renderForEffects$3(fragment, effectors, ...scope));
		}
		static {
			this.OPEN_RE = /<([a-zA-z][$a-zA-Z0-9.-]*)\s+[^>]*$/;
		}
		static {
			this.IN_TAG_RE = /^(\s+|[^<>]*|"[^"]*")*$/;
		}
		static {
			this.ATTR_RE = /([$.]?[a-zA-Z][$a-zA-Z0-9.-]*)=\s*$/;
		}
		static {
			this.CLOSE_RE = /[/]?>[^<]*$/;
		}
		classifyPlace(i, template) {
			let tagOpen = null;
			for (let prev = i; prev >= 0; prev--) {
				if (template[prev].match(TemplateParser.CLOSE_RE)) break;
				tagOpen = template[prev].match(TemplateParser.OPEN_RE);
				if (tagOpen) break;
				if (!template[prev].match(TemplateParser.IN_TAG_RE)) break;
			}
			if (tagOpen) {
				const tagAttr = template[i].match(TemplateParser.ATTR_RE);
				if (tagAttr) return {
					kind: "attr value",
					nodeLabel: `node${i}`,
					tagName: tagOpen[1],
					attrName: tagAttr[1]
				};
				return {
					kind: "tag content",
					nodeLabel: `node${i}`,
					tagName: tagOpen[1]
				};
			}
			return {
				kind: "element content",
				nodeLabel: `node${i}`
			};
		}
		tryReplacements(place, param) {
			const replacements = this.plugins;
			for (let i = 0; i < replacements.length; i++) {
				const sub = replacements[i];
				if (place.kind === sub.place && checkType$3(param, sub)) return sub.mutator(place, param);
			}
		}
	};
	function checkType$3(param, sub) {
		if (typeof sub.types === "function") return sub.types(param, sub);
		return sub.types.includes(typeof param);
	}
	var Mutation$3 = class {
		constructor(place) {
			this.place = place;
		}
		apply(_site, _fragment) {
			throw "abstract method 'apply' called";
		}
	};
	var ElementContentMutation$3 = class extends Mutation$3 {
		constructor(place, content) {
			super(place);
			this.content = content;
		}
		apply(site, fragment) {
			(site.parentNode || fragment).replaceChild(this.content, site);
		}
	};
	var AttributeValueMutation$3 = class extends Mutation$3 {
		constructor(place, text) {
			super(place);
			this.text = text;
			this.name = place.attrName;
		}
		apply(site) {
			site.setAttribute(this.name, this.text);
		}
	};
	var ElementContentEffect$3 = class extends Mutation$3 {
		constructor(place, fn) {
			super(place);
			this.fn = fn;
		}
		apply(_site, _fragment) {
			const key = this.place.nodeLabel;
			return (site, fragment, ...scope) => {
				const start = new Comment(` <<< ${key} `);
				const end = new Comment(` >>> ${key} `);
				const placeholder = new DocumentFragment();
				placeholder.replaceChildren(start, end);
				(site.parentNode || fragment).replaceChild(placeholder, site);
				createEffect$3((...args) => {
					replaceElementContent$3(this.fn(...args), start, end);
				}, ...scope);
			};
		}
	};
	function replaceElementContent$3(value, start, end) {
		const parent = start.parentNode;
		if (!parent) throw new Error("No parent for placeholder");
		const valueToNode = (v) => {
			if (Array.isArray(v)) {
				const frag = new DocumentFragment();
				const nodes = v.map(valueToNode);
				frag.replaceChildren(...nodes);
				return frag;
			} else if (v instanceof Node) return v;
			else return new Text(v?.toString() || "");
		};
		const node = valueToNode(value);
		let p = start.nextSibling;
		while (p && p !== end) {
			const old = p;
			p = p.nextSibling;
			parent.removeChild(old);
		}
		if (node) parent.insertBefore(node, end);
	}
	var AttributeValueEffect$3 = class extends Mutation$3 {
		constructor(place, fn) {
			super(place);
			this.fn = fn;
			this.name = place.attrName;
		}
		apply(_site, _fragment) {
			return (site, _, ...scope) => createEffect$3((...args) => {
				replaceAttributeValue$3(this.fn(...args), site, this.name);
			}, ...scope);
		}
	};
	function replaceAttributeValue$3(value, site, attrName) {
		const special = attrName.match(/^([.$])(.+)$/);
		if (special) {
			const [_, pre, name] = special;
			switch (pre) {
				case ".":
					site[name] = value;
					break;
				case "$":
					if ("viewModel" in site && site.viewModel instanceof Context$3) site.viewModel.set(name, value);
					break;
			}
		} else switch (typeof value) {
			case "string":
				site.setAttribute(attrName, value);
				break;
			case "undefined":
			case "object":
			case "boolean":
				if (value) site.setAttribute(attrName, attrName);
				else site.removeAttribute(attrName);
				break;
			default: site.setAttribute(attrName, value?.toString());
		}
	}
	var TagReferenceEffect$3 = class extends Mutation$3 {
		constructor(place, fn) {
			super(place);
			this.fn = fn;
		}
		apply(_site, _fragment) {
			return (site, _, ...scope) => createEffect$3((...args) => {
				const value = this.fn(...args);
				if (typeof value === "function") value(site);
			}, ...scope);
		}
	};
	new TemplateParser$3().use([
		{
			place: "element content",
			types: [
				"string",
				"number",
				"bigint",
				"symbol",
				"boolean"
			],
			mutator: (place, value) => new ElementContentMutation$3(place, new Text(value?.toString() || ""))
		},
		{
			place: "attr value",
			types: [
				"string",
				"number",
				"bigint",
				"symbol"
			],
			mutator: (place, value) => new AttributeValueMutation$3(place, value?.toString() || "")
		},
		{
			place: "element content",
			types: (param) => param instanceof Node,
			mutator: (place, value) => new ElementContentMutation$3(place, value)
		},
		{
			place: "element content",
			types: (param) => Array.isArray(param),
			mutator: (place, value) => {
				const fragment = new DocumentFragment();
				const nodes = value.map((v) => v instanceof Node ? v : new Text(v?.toString() || ""));
				fragment.append(...nodes);
				return new ElementContentMutation$3(place, fragment);
			}
		},
		{
			place: "element content",
			types: ["function"],
			mutator: (place, param) => new ElementContentEffect$3(place, param)
		},
		{
			place: "attr value",
			types: ["function"],
			mutator: (place, param) => new AttributeValueEffect$3(place, param)
		},
		{
			place: "tag content",
			types: ["function"],
			mutator: (place, param) => new TagReferenceEffect$3(place, param)
		}
	]);
	var Provider$1 = class Provider extends HTMLElement {
		static {
			this.DISCOVERY_EVENT = "un-provider:discover";
		}
		static {
			this.REGISTRY_EVENT = "un-provider:register";
		}
		static {
			this.CHANGE_EVENT = "un-provider:change";
		}
		static {
			document.addEventListener(Provider.DISCOVERY_EVENT, (event) => {
				const [contextLabel, respondFn] = event.detail;
				const provider = registeredProvider$1(contextLabel);
				if (provider) respondFn(provider);
			});
			document.addEventListener(Provider.REGISTRY_EVENT, (event) => {
				const [contextLabel, provider] = event.detail;
				registerProvider$1(contextLabel, provider);
			});
		}
		constructor(init, label) {
			super();
			this.contextLabel = label;
			this.context = new Context$3(init);
			this.context.setHost(this, Provider.CHANGE_EVENT);
			this.addEventListener(Provider.DISCOVERY_EVENT, (event) => {
				const [contextLabel, respondFn] = event.detail;
				if (contextLabel === this.contextLabel) {
					event.stopPropagation();
					respondFn(this);
				}
			});
			const registryEvent = new CustomEvent(Provider.REGISTRY_EVENT, {
				bubbles: true,
				composed: true,
				detail: [this.contextLabel, this]
			});
			this.dispatchEvent(registryEvent);
		}
		attach(observer) {
			this.addEventListener(Provider.CHANGE_EVENT, observer);
			return this.context.toObject();
		}
		detach(observer) {
			this.removeEventListener(Provider.CHANGE_EVENT, observer);
		}
	};
	function discover$1(observer, contextLabel) {
		return new Promise((resolve, reject) => {
			const discoveryEvent = new CustomEvent(Provider$1.DISCOVERY_EVENT, {
				bubbles: true,
				composed: true,
				detail: [contextLabel, (provider) => provider ? resolve(provider) : reject()]
			});
			if (observer.isConnected) observer.dispatchEvent(discoveryEvent);
			else document.dispatchEvent(discoveryEvent);
		});
	}
	var registry$1 = {};
	function registerProvider$1(label, provider) {
		registry$1[label] = provider;
	}
	function registeredProvider$1(label) {
		return registry$1[label];
	}
	var Observer$1 = class {
		constructor(contextLabel) {
			this.contextLabel = contextLabel;
		}
		observe(from, fn) {
			return new Promise((resolve, reject) => {
				if (this.provider) resolve(this.attachObserver(fn));
				else discover$1(from, this.contextLabel).then((provider) => {
					this.provider = provider;
					resolve(this.attachObserver(fn));
				}).catch((err) => reject(err));
			});
		}
		attachObserver(fn) {
			const init = this.provider.attach((ev) => {
				const { property, value } = ev.detail;
				const effect = new DirectEffect$1(fn, {
					property,
					value
				});
				console.log("📡Signal received:", property, value);
				effect.execute();
			});
			console.log("👀 Initial observation:", init);
			return init;
		}
	};
	var Service$1 = class {
		constructor(update, context, eventType = "service:message", autostart = true) {
			this._pending = [];
			this._context = context;
			this._update = update;
			this._eventType = eventType;
			this._running = autostart;
		}
		attach(host) {
			host.addEventListener(this._eventType, (ev) => {
				ev.stopPropagation();
				const message = ev.detail;
				this.consume(message);
			});
		}
		start() {
			if (!this._running) {
				console.log(`Starting ${this._eventType} service`);
				this._running = true;
				this._pending.forEach((msg) => this.process(msg));
			}
		}
		consume(message) {
			if (message.length !== 0) {
				const trueMsg = message;
				if (this._running) this.process(trueMsg);
				else {
					console.log(`📥 Queueing ${this._eventType} message`, message);
					this._pending.push(trueMsg);
				}
			}
		}
		process(message) {
			console.log(`📤 Processing ${this._eventType} message`, message);
			const next = this._update(message, this._context.toObject());
			console.log("⏭️ next state: ", next);
			if (!Array.isArray(next)) this._context.update(next);
			else {
				const [now, ...later] = next;
				later.forEach((promise) => promise.then((command) => this.consume(command)));
				this._context.update(now);
			}
		}
	};
	function fromService$1(target, contextLabel) {
		return new FromService$1(target, contextLabel);
	}
	var FromService$1 = class {
		constructor(client, contextLabel) {
			this.client = client;
			this.observer = new Observer$1(contextLabel);
		}
		start(fn) {
			return this.observer.observe(this.client, (s) => {
				fn(s.property, s.value);
			});
		}
	};
	//#endregion
	//#region src/history.ts
	var history_exports = /* @__PURE__ */ __exportAll$3({
		CONTEXT_DEFAULT: () => HISTORY_CONTEXT_DEFAULT,
		Provider: () => HistoryProvider,
		Service: () => HistoryService,
		dispatch: () => dispatch$2
	});
	var HISTORY_CONTEXT_DEFAULT = "context:history";
	var cloneCurrentLocation = () => Object.assign({}, document.location);
	var HistoryService = class HistoryService extends Service$1 {
		static {
			this.EVENT_TYPE = "history:message";
		}
		constructor(context) {
			super((msg, model) => this.update(msg, model), context, HistoryService.EVENT_TYPE);
		}
		update(message, model) {
			switch (message[0]) {
				case "history/navigate": {
					const { href, state } = message[1];
					history.pushState(state, "", href);
					return {
						...model,
						location: cloneCurrentLocation(),
						state: history.state
					};
				}
				case "history/redirect": {
					const { href, state } = message[1];
					history.replaceState(state, "", href);
					return {
						...model,
						location: cloneCurrentLocation(),
						state: history.state
					};
				}
			}
			return model;
		}
	};
	var HistoryProvider = class extends Provider$1 {
		constructor() {
			super({
				location: cloneCurrentLocation(),
				state: {}
			}, HISTORY_CONTEXT_DEFAULT);
			this.addEventListener("click", (event) => {
				const linkTarget = originalLinkTarget(event);
				if (linkTarget) {
					const url = new URL(linkTarget.href);
					const location = this.context.get("location");
					if (location && url.origin === location.origin && url.pathname.startsWith(this.base || "/")) {
						event.preventDefault();
						dispatch$2(linkTarget, "history/navigate", { href: url.pathname + url.search });
					}
				}
			});
			window.addEventListener("popstate", (event) => {
				this.context.update({
					location: cloneCurrentLocation(),
					state: event.state
				});
			});
		}
		get base() {
			return this.getAttribute("base") || void 0;
		}
		connectedCallback() {
			new HistoryService(this.context).attach(this);
		}
		attributeChangedCallback() {}
	};
	function originalLinkTarget(event) {
		const current = event.currentTarget;
		const isLink = (el) => el.tagName == "A" && el.href;
		if (event.button !== 0) return void 0;
		if (event.composed) {
			const target = event.composedPath().find(isLink);
			return target ? target : void 0;
		} else {
			for (let target = event.target; target; target === current || target.parentElement) if (isLink(target)) return target;
			return;
		}
	}
	var dispatch$2 = message_exports$1.dispatcher(HistoryService.EVENT_TYPE);
	//#endregion
	//#region ../../node_modules/route-parser/lib/route/compiled-grammar.js
	var require_compiled_grammar = /* @__PURE__ */ __commonJSMin(((exports) => {
		var parser = (function() {
			var o = function(k, v, o, l) {
				for (o = o || {}, l = k.length; l--; o[k[l]] = v);
				return o;
			}, $V0 = [1, 9], $V1 = [1, 10], $V2 = [1, 11], $V3 = [1, 12], $V4 = [
				5,
				11,
				12,
				13,
				14,
				15
			];
			var parser = {
				trace: function trace() {},
				yy: {},
				symbols_: {
					"error": 2,
					"root": 3,
					"expressions": 4,
					"EOF": 5,
					"expression": 6,
					"optional": 7,
					"literal": 8,
					"splat": 9,
					"param": 10,
					"(": 11,
					")": 12,
					"LITERAL": 13,
					"SPLAT": 14,
					"PARAM": 15,
					"$accept": 0,
					"$end": 1
				},
				terminals_: {
					2: "error",
					5: "EOF",
					11: "(",
					12: ")",
					13: "LITERAL",
					14: "SPLAT",
					15: "PARAM"
				},
				productions_: [
					0,
					[3, 2],
					[3, 1],
					[4, 2],
					[4, 1],
					[6, 1],
					[6, 1],
					[6, 1],
					[6, 1],
					[7, 3],
					[8, 1],
					[9, 1],
					[10, 1]
				],
				performAction: function anonymous(yytext, yyleng, yylineno, yy, yystate, $$, _$) {
					var $0 = $$.length - 1;
					switch (yystate) {
						case 1: return new yy.Root({}, [$$[$0 - 1]]);
						case 2: return new yy.Root({}, [new yy.Literal({ value: "" })]);
						case 3:
							this.$ = new yy.Concat({}, [$$[$0 - 1], $$[$0]]);
							break;
						case 4:
						case 5:
							this.$ = $$[$0];
							break;
						case 6:
							this.$ = new yy.Literal({ value: $$[$0] });
							break;
						case 7:
							this.$ = new yy.Splat({ name: $$[$0] });
							break;
						case 8:
							this.$ = new yy.Param({ name: $$[$0] });
							break;
						case 9:
							this.$ = new yy.Optional({}, [$$[$0 - 1]]);
							break;
						case 10:
							this.$ = yytext;
							break;
						case 11:
						case 12:
							this.$ = yytext.slice(1);
							break;
					}
				},
				table: [
					{
						3: 1,
						4: 2,
						5: [1, 3],
						6: 4,
						7: 5,
						8: 6,
						9: 7,
						10: 8,
						11: $V0,
						13: $V1,
						14: $V2,
						15: $V3
					},
					{ 1: [3] },
					{
						5: [1, 13],
						6: 14,
						7: 5,
						8: 6,
						9: 7,
						10: 8,
						11: $V0,
						13: $V1,
						14: $V2,
						15: $V3
					},
					{ 1: [2, 2] },
					o($V4, [2, 4]),
					o($V4, [2, 5]),
					o($V4, [2, 6]),
					o($V4, [2, 7]),
					o($V4, [2, 8]),
					{
						4: 15,
						6: 4,
						7: 5,
						8: 6,
						9: 7,
						10: 8,
						11: $V0,
						13: $V1,
						14: $V2,
						15: $V3
					},
					o($V4, [2, 10]),
					o($V4, [2, 11]),
					o($V4, [2, 12]),
					{ 1: [2, 1] },
					o($V4, [2, 3]),
					{
						6: 14,
						7: 5,
						8: 6,
						9: 7,
						10: 8,
						11: $V0,
						12: [1, 16],
						13: $V1,
						14: $V2,
						15: $V3
					},
					o($V4, [2, 9])
				],
				defaultActions: {
					3: [2, 2],
					13: [2, 1]
				},
				parseError: function parseError(str, hash) {
					if (hash.recoverable) this.trace(str);
					else {
						function _parseError(msg, hash) {
							this.message = msg;
							this.hash = hash;
						}
						_parseError.prototype = Error;
						throw new _parseError(str, hash);
					}
				},
				parse: function parse(input) {
					var self = this, stack = [0], vstack = [null], lstack = [], table = this.table, yytext = "", yylineno = 0, yyleng = 0, recovering = 0, TERROR = 2, EOF = 1;
					var args = lstack.slice.call(arguments, 1);
					var lexer = Object.create(this.lexer);
					var sharedState = { yy: {} };
					for (var k in this.yy) if (Object.prototype.hasOwnProperty.call(this.yy, k)) sharedState.yy[k] = this.yy[k];
					lexer.setInput(input, sharedState.yy);
					sharedState.yy.lexer = lexer;
					sharedState.yy.parser = this;
					if (typeof lexer.yylloc == "undefined") lexer.yylloc = {};
					var yyloc = lexer.yylloc;
					lstack.push(yyloc);
					var ranges = lexer.options && lexer.options.ranges;
					if (typeof sharedState.yy.parseError === "function") this.parseError = sharedState.yy.parseError;
					else this.parseError = Object.getPrototypeOf(this).parseError;
					_token_stack: var lex = function() {
						var token = lexer.lex() || EOF;
						if (typeof token !== "number") token = self.symbols_[token] || token;
						return token;
					};
					var symbol, preErrorSymbol, state, action, r, yyval = {}, p, len, newState, expected;
					while (true) {
						state = stack[stack.length - 1];
						if (this.defaultActions[state]) action = this.defaultActions[state];
						else {
							if (symbol === null || typeof symbol == "undefined") symbol = lex();
							action = table[state] && table[state][symbol];
						}
						if (typeof action === "undefined" || !action.length || !action[0]) {
							var errStr = "";
							expected = [];
							for (p in table[state]) if (this.terminals_[p] && p > TERROR) expected.push("'" + this.terminals_[p] + "'");
							if (lexer.showPosition) errStr = "Parse error on line " + (yylineno + 1) + ":\n" + lexer.showPosition() + "\nExpecting " + expected.join(", ") + ", got '" + (this.terminals_[symbol] || symbol) + "'";
							else errStr = "Parse error on line " + (yylineno + 1) + ": Unexpected " + (symbol == EOF ? "end of input" : "'" + (this.terminals_[symbol] || symbol) + "'");
							this.parseError(errStr, {
								text: lexer.match,
								token: this.terminals_[symbol] || symbol,
								line: lexer.yylineno,
								loc: yyloc,
								expected
							});
						}
						if (action[0] instanceof Array && action.length > 1) throw new Error("Parse Error: multiple actions possible at state: " + state + ", token: " + symbol);
						switch (action[0]) {
							case 1:
								stack.push(symbol);
								vstack.push(lexer.yytext);
								lstack.push(lexer.yylloc);
								stack.push(action[1]);
								symbol = null;
								if (!preErrorSymbol) {
									yyleng = lexer.yyleng;
									yytext = lexer.yytext;
									yylineno = lexer.yylineno;
									yyloc = lexer.yylloc;
									if (recovering > 0) recovering--;
								} else {
									symbol = preErrorSymbol;
									preErrorSymbol = null;
								}
								break;
							case 2:
								len = this.productions_[action[1]][1];
								yyval.$ = vstack[vstack.length - len];
								yyval._$ = {
									first_line: lstack[lstack.length - (len || 1)].first_line,
									last_line: lstack[lstack.length - 1].last_line,
									first_column: lstack[lstack.length - (len || 1)].first_column,
									last_column: lstack[lstack.length - 1].last_column
								};
								if (ranges) yyval._$.range = [lstack[lstack.length - (len || 1)].range[0], lstack[lstack.length - 1].range[1]];
								r = this.performAction.apply(yyval, [
									yytext,
									yyleng,
									yylineno,
									sharedState.yy,
									action[1],
									vstack,
									lstack
								].concat(args));
								if (typeof r !== "undefined") return r;
								if (len) {
									stack = stack.slice(0, -1 * len * 2);
									vstack = vstack.slice(0, -1 * len);
									lstack = lstack.slice(0, -1 * len);
								}
								stack.push(this.productions_[action[1]][0]);
								vstack.push(yyval.$);
								lstack.push(yyval._$);
								newState = table[stack[stack.length - 2]][stack[stack.length - 1]];
								stack.push(newState);
								break;
							case 3: return true;
						}
					}
					return true;
				}
			};
			parser.lexer = (function() {
				return {
					EOF: 1,
					parseError: function parseError(str, hash) {
						if (this.yy.parser) this.yy.parser.parseError(str, hash);
						else throw new Error(str);
					},
					setInput: function(input, yy) {
						this.yy = yy || this.yy || {};
						this._input = input;
						this._more = this._backtrack = this.done = false;
						this.yylineno = this.yyleng = 0;
						this.yytext = this.matched = this.match = "";
						this.conditionStack = ["INITIAL"];
						this.yylloc = {
							first_line: 1,
							first_column: 0,
							last_line: 1,
							last_column: 0
						};
						if (this.options.ranges) this.yylloc.range = [0, 0];
						this.offset = 0;
						return this;
					},
					input: function() {
						var ch = this._input[0];
						this.yytext += ch;
						this.yyleng++;
						this.offset++;
						this.match += ch;
						this.matched += ch;
						if (ch.match(/(?:\r\n?|\n).*/g)) {
							this.yylineno++;
							this.yylloc.last_line++;
						} else this.yylloc.last_column++;
						if (this.options.ranges) this.yylloc.range[1]++;
						this._input = this._input.slice(1);
						return ch;
					},
					unput: function(ch) {
						var len = ch.length;
						var lines = ch.split(/(?:\r\n?|\n)/g);
						this._input = ch + this._input;
						this.yytext = this.yytext.substr(0, this.yytext.length - len);
						this.offset -= len;
						var oldLines = this.match.split(/(?:\r\n?|\n)/g);
						this.match = this.match.substr(0, this.match.length - 1);
						this.matched = this.matched.substr(0, this.matched.length - 1);
						if (lines.length - 1) this.yylineno -= lines.length - 1;
						var r = this.yylloc.range;
						this.yylloc = {
							first_line: this.yylloc.first_line,
							last_line: this.yylineno + 1,
							first_column: this.yylloc.first_column,
							last_column: lines ? (lines.length === oldLines.length ? this.yylloc.first_column : 0) + oldLines[oldLines.length - lines.length].length - lines[0].length : this.yylloc.first_column - len
						};
						if (this.options.ranges) this.yylloc.range = [r[0], r[0] + this.yyleng - len];
						this.yyleng = this.yytext.length;
						return this;
					},
					more: function() {
						this._more = true;
						return this;
					},
					reject: function() {
						if (this.options.backtrack_lexer) this._backtrack = true;
						else return this.parseError("Lexical error on line " + (this.yylineno + 1) + ". You can only invoke reject() in the lexer when the lexer is of the backtracking persuasion (options.backtrack_lexer = true).\n" + this.showPosition(), {
							text: "",
							token: null,
							line: this.yylineno
						});
						return this;
					},
					less: function(n) {
						this.unput(this.match.slice(n));
					},
					pastInput: function() {
						var past = this.matched.substr(0, this.matched.length - this.match.length);
						return (past.length > 20 ? "..." : "") + past.substr(-20).replace(/\n/g, "");
					},
					upcomingInput: function() {
						var next = this.match;
						if (next.length < 20) next += this._input.substr(0, 20 - next.length);
						return (next.substr(0, 20) + (next.length > 20 ? "..." : "")).replace(/\n/g, "");
					},
					showPosition: function() {
						var pre = this.pastInput();
						var c = new Array(pre.length + 1).join("-");
						return pre + this.upcomingInput() + "\n" + c + "^";
					},
					test_match: function(match, indexed_rule) {
						var token, lines, backup;
						if (this.options.backtrack_lexer) {
							backup = {
								yylineno: this.yylineno,
								yylloc: {
									first_line: this.yylloc.first_line,
									last_line: this.last_line,
									first_column: this.yylloc.first_column,
									last_column: this.yylloc.last_column
								},
								yytext: this.yytext,
								match: this.match,
								matches: this.matches,
								matched: this.matched,
								yyleng: this.yyleng,
								offset: this.offset,
								_more: this._more,
								_input: this._input,
								yy: this.yy,
								conditionStack: this.conditionStack.slice(0),
								done: this.done
							};
							if (this.options.ranges) backup.yylloc.range = this.yylloc.range.slice(0);
						}
						lines = match[0].match(/(?:\r\n?|\n).*/g);
						if (lines) this.yylineno += lines.length;
						this.yylloc = {
							first_line: this.yylloc.last_line,
							last_line: this.yylineno + 1,
							first_column: this.yylloc.last_column,
							last_column: lines ? lines[lines.length - 1].length - lines[lines.length - 1].match(/\r?\n?/)[0].length : this.yylloc.last_column + match[0].length
						};
						this.yytext += match[0];
						this.match += match[0];
						this.matches = match;
						this.yyleng = this.yytext.length;
						if (this.options.ranges) this.yylloc.range = [this.offset, this.offset += this.yyleng];
						this._more = false;
						this._backtrack = false;
						this._input = this._input.slice(match[0].length);
						this.matched += match[0];
						token = this.performAction.call(this, this.yy, this, indexed_rule, this.conditionStack[this.conditionStack.length - 1]);
						if (this.done && this._input) this.done = false;
						if (token) return token;
						else if (this._backtrack) {
							for (var k in backup) this[k] = backup[k];
							return false;
						}
						return false;
					},
					next: function() {
						if (this.done) return this.EOF;
						if (!this._input) this.done = true;
						var token, match, tempMatch, index;
						if (!this._more) {
							this.yytext = "";
							this.match = "";
						}
						var rules = this._currentRules();
						for (var i = 0; i < rules.length; i++) {
							tempMatch = this._input.match(this.rules[rules[i]]);
							if (tempMatch && (!match || tempMatch[0].length > match[0].length)) {
								match = tempMatch;
								index = i;
								if (this.options.backtrack_lexer) {
									token = this.test_match(tempMatch, rules[i]);
									if (token !== false) return token;
									else if (this._backtrack) {
										match = false;
										continue;
									} else return false;
								} else if (!this.options.flex) break;
							}
						}
						if (match) {
							token = this.test_match(match, rules[index]);
							if (token !== false) return token;
							return false;
						}
						if (this._input === "") return this.EOF;
						else return this.parseError("Lexical error on line " + (this.yylineno + 1) + ". Unrecognized text.\n" + this.showPosition(), {
							text: "",
							token: null,
							line: this.yylineno
						});
					},
					lex: function lex() {
						var r = this.next();
						if (r) return r;
						else return this.lex();
					},
					begin: function begin(condition) {
						this.conditionStack.push(condition);
					},
					popState: function popState() {
						if (this.conditionStack.length - 1 > 0) return this.conditionStack.pop();
						else return this.conditionStack[0];
					},
					_currentRules: function _currentRules() {
						if (this.conditionStack.length && this.conditionStack[this.conditionStack.length - 1]) return this.conditions[this.conditionStack[this.conditionStack.length - 1]].rules;
						else return this.conditions["INITIAL"].rules;
					},
					topState: function topState(n) {
						n = this.conditionStack.length - 1 - Math.abs(n || 0);
						if (n >= 0) return this.conditionStack[n];
						else return "INITIAL";
					},
					pushState: function pushState(condition) {
						this.begin(condition);
					},
					stateStackSize: function stateStackSize() {
						return this.conditionStack.length;
					},
					options: {},
					performAction: function anonymous(yy, yy_, $avoiding_name_collisions, YY_START) {
						switch ($avoiding_name_collisions) {
							case 0: return "(";
							case 1: return ")";
							case 2: return "SPLAT";
							case 3: return "PARAM";
							case 4: return "LITERAL";
							case 5: return "LITERAL";
							case 6: return "EOF";
						}
					},
					rules: [
						/^(?:\()/,
						/^(?:\))/,
						/^(?:\*+\w+)/,
						/^(?::+\w+)/,
						/^(?:[\w%\-~\n]+)/,
						/^(?:.)/,
						/^(?:$)/
					],
					conditions: { "INITIAL": {
						"rules": [
							0,
							1,
							2,
							3,
							4,
							5,
							6
						],
						"inclusive": true
					} }
				};
			})();
			function Parser() {
				this.yy = {};
			}
			Parser.prototype = parser;
			parser.Parser = Parser;
			return new Parser();
		})();
		if (typeof require !== "undefined" && typeof exports !== "undefined") {
			exports.parser = parser;
			exports.Parser = parser.Parser;
			exports.parse = function() {
				return parser.parse.apply(parser, arguments);
			};
		}
	}));
	//#endregion
	//#region ../../node_modules/route-parser/lib/route/nodes.js
	var require_nodes = /* @__PURE__ */ __commonJSMin(((exports, module) => {
		/** @module route/nodes */
		/**
		* Create a node for use with the parser, giving it a constructor that takes
		* props, children, and returns an object with props, children, and a
		* displayName.
		* @param  {String} displayName The display name for the node
		* @return {{displayName: string, props: Object, children: Array}}
		*/
		function createNode(displayName) {
			return function(props, children) {
				return {
					displayName,
					props,
					children: children || []
				};
			};
		}
		module.exports = {
			Root: createNode("Root"),
			Concat: createNode("Concat"),
			Literal: createNode("Literal"),
			Splat: createNode("Splat"),
			Param: createNode("Param"),
			Optional: createNode("Optional")
		};
	}));
	//#endregion
	//#region ../../node_modules/route-parser/lib/route/parser.js
	/**
	* @module route/parser
	*/
	var require_parser = /* @__PURE__ */ __commonJSMin(((exports, module) => {
		/** Wrap the compiled parser with the context to create node objects */
		var parser = require_compiled_grammar().parser;
		parser.yy = require_nodes();
		module.exports = parser;
	}));
	//#endregion
	//#region ../../node_modules/route-parser/lib/route/visitors/create_visitor.js
	var require_create_visitor = /* @__PURE__ */ __commonJSMin(((exports, module) => {
		/**
		* @module route/visitors/create_visitor
		*/
		var nodeTypes = Object.keys(require_nodes());
		/**
		* Helper for creating visitors. Take an object of node name to handler
		* mappings, returns an object with a "visit" method that can be called
		* @param  {Object.<string,function(node,context)>} handlers A mapping of node
		* type to visitor functions
		* @return {{visit: function(node,context)}}  A visitor object with a "visit"
		* method that can be called on a node with a context
		*/
		function createVisitor(handlers) {
			nodeTypes.forEach(function(nodeType) {
				if (typeof handlers[nodeType] === "undefined") throw new Error("No handler defined for " + nodeType.displayName);
			});
			return {
				/**
				* Call the given handler for this node type
				* @param  {Object} node    the AST node
				* @param  {Object} context context to pass through to handlers
				* @return {Object}
				*/
				visit: function(node, context) {
					return this.handlers[node.displayName].call(this, node, context);
				},
				handlers
			};
		}
		module.exports = createVisitor;
	}));
	//#endregion
	//#region ../../node_modules/route-parser/lib/route/visitors/regexp.js
	var require_regexp = /* @__PURE__ */ __commonJSMin(((exports, module) => {
		var createVisitor = require_create_visitor(), escapeRegExp = /[\-{}\[\]+?.,\\\^$|#\s]/g;
		/**
		* @class
		* @private
		*/
		function Matcher(options) {
			this.captures = options.captures;
			this.re = options.re;
		}
		/**
		* Try matching a path against the generated regular expression
		* @param  {String} path The path to try to match
		* @return {Object|false}      matched parameters or false
		*/
		Matcher.prototype.match = function(path) {
			var match = this.re.exec(path), matchParams = {};
			if (!match) return;
			this.captures.forEach(function(capture, i) {
				if (typeof match[i + 1] === "undefined") matchParams[capture] = void 0;
				else matchParams[capture] = decodeURIComponent(match[i + 1]);
			});
			return matchParams;
		};
		module.exports = createVisitor({
			"Concat": function(node) {
				return node.children.reduce(function(memo, child) {
					var childResult = this.visit(child);
					return {
						re: memo.re + childResult.re,
						captures: memo.captures.concat(childResult.captures)
					};
				}.bind(this), {
					re: "",
					captures: []
				});
			},
			"Literal": function(node) {
				return {
					re: node.props.value.replace(escapeRegExp, "\\$&"),
					captures: []
				};
			},
			"Splat": function(node) {
				return {
					re: "([^?]*?)",
					captures: [node.props.name]
				};
			},
			"Param": function(node) {
				return {
					re: "([^\\/\\?]+)",
					captures: [node.props.name]
				};
			},
			"Optional": function(node) {
				var child = this.visit(node.children[0]);
				return {
					re: "(?:" + child.re + ")?",
					captures: child.captures
				};
			},
			"Root": function(node) {
				var childResult = this.visit(node.children[0]);
				return new Matcher({
					re: new RegExp("^" + childResult.re + "(?=\\?|$)"),
					captures: childResult.captures
				});
			}
		});
	}));
	//#endregion
	//#region ../../node_modules/route-parser/lib/route/visitors/reverse.js
	var require_reverse = /* @__PURE__ */ __commonJSMin(((exports, module) => {
		module.exports = require_create_visitor()({
			"Concat": function(node, context) {
				var childResults = node.children.map(function(child) {
					return this.visit(child, context);
				}.bind(this));
				if (childResults.some(function(c) {
					return c === false;
				})) return false;
				else return childResults.join("");
			},
			"Literal": function(node) {
				return decodeURI(node.props.value);
			},
			"Splat": function(node, context) {
				if (context[node.props.name]) return context[node.props.name];
				else return false;
			},
			"Param": function(node, context) {
				if (context[node.props.name]) return context[node.props.name];
				else return false;
			},
			"Optional": function(node, context) {
				var childResult = this.visit(node.children[0], context);
				if (childResult) return childResult;
				else return "";
			},
			"Root": function(node, context) {
				context = context || {};
				var childResult = this.visit(node.children[0], context);
				if (!childResult) return false;
				return encodeURI(childResult);
			}
		});
	}));
	//#endregion
	//#region ../../node_modules/route-parser/lib/route.js
	var require_route = /* @__PURE__ */ __commonJSMin(((exports, module) => {
		var Parser = require_parser(), RegexpVisitor = require_regexp(), ReverseVisitor = require_reverse();
		Route.prototype = Object.create(null);
		/**
		* Match a path against this route, returning the matched parameters if
		* it matches, false if not.
		* @example
		* var route = new Route('/this/is/my/route')
		* route.match('/this/is/my/route') // -> {}
		* @example
		* var route = new Route('/:one/:two')
		* route.match('/foo/bar/') // -> {one: 'foo', two: 'bar'}
		* @param  {string} path the path to match this route against
		* @return {(Object.<string,string>|false)} A map of the matched route
		* parameters, or false if matching failed
		*/
		Route.prototype.match = function(path) {
			var matched = RegexpVisitor.visit(this.ast).match(path);
			return matched ? matched : false;
		};
		/**
		* Reverse a route specification to a path, returning false if it can't be
		* fulfilled
		* @example
		* var route = new Route('/:one/:two')
		* route.reverse({one: 'foo', two: 'bar'}) -> '/foo/bar'
		* @param  {Object} params The parameters to fill in
		* @return {(String|false)} The filled in path
		*/
		Route.prototype.reverse = function(params) {
			return ReverseVisitor.visit(this.ast, params);
		};
		/**
		* Represents a route
		* @example
		* var route = Route('/:foo/:bar');
		* @example
		* var route = Route('/:foo/:bar');
		* @param {string} spec -  the string specification of the route.
		*     use :param for single portion captures, *param for splat style captures,
		*     and () for optional route branches
		* @constructor
		*/
		function Route(spec) {
			var route;
			if (this) route = this;
			else route = Object.create(Route.prototype);
			if (typeof spec === "undefined") throw new Error("A route spec is required");
			route.spec = spec;
			route.ast = Parser.parse(spec);
			return route;
		}
		module.exports = Route;
	}));
	//#endregion
	//#region ../html/dist/html.js
	var import_route_parser = /* @__PURE__ */ __toESM((/* @__PURE__ */ __commonJSMin(((exports, module) => {
		module.exports = require_route();
	})))(), 1);
	var SignalEvent$2 = class extends CustomEvent {
		constructor(eventType, signal) {
			super(eventType, {
				bubbles: true,
				cancelable: true,
				composed: true,
				detail: signal
			});
		}
	};
	function createEffect$2(fn, ...scope) {
		const effect = { execute() {
			console.log("▶️ Executing created effect:", scope, fn);
			fn(...scope.map((cx) => cx.open(effect)));
			scope.forEach((cx) => cx.close());
		} };
		effect.execute();
	}
	var Scheduler$2 = class Scheduler {
		constructor() {
			this.signals = /* @__PURE__ */ new WeakMap();
			this.scheduled = /* @__PURE__ */ new WeakSet();
		}
		static {
			this.scheduler = new Scheduler();
		}
		subscribe(scope, key, effect) {
			let signals = this.signals.get(scope);
			if (!signals) {
				signals = /* @__PURE__ */ new Map();
				this.signals.set(scope, signals);
			}
			let signal = signals.get(key);
			if (!signal) {
				signal = /* @__PURE__ */ new Set();
				signals.set(key, signal);
			}
			signal.add(effect);
		}
		scheduleEffects(scope, key) {
			const signals = this.signals.get(scope);
			if (!signals) return;
			const signal = signals.get(key);
			if (signal) {
				for (const effect of signal) if (!this.scheduled.has(effect)) {
					this.scheduled.add(effect);
					setTimeout(() => {
						this.scheduled.delete(effect);
						effect.execute();
					});
				}
			}
		}
	};
	var EffectsManager$2 = class {
		constructor() {
			this.running = [];
			this.eventType = "un-effect:change";
		}
		isRunning() {
			return this.running.length > 0;
		}
		push(effect) {
			this.running.push(effect);
		}
		pop() {
			this.running.pop();
		}
		current() {
			const len = this.running.length;
			return len ? this.running[len - 1] : void 0;
		}
		subscribe(key, scope) {
			const current = this.current();
			if (current) Scheduler$2.scheduler.subscribe(scope, key, current);
		}
		runEffects(key, scope) {
			console.log("⚙️ Running effects for signal", key, scope);
			Scheduler$2.scheduler.scheduleEffects(scope, key);
			if (this.host) {
				const evt = new SignalEvent$2(this.eventType, {
					property: key,
					value: scope[key]
				});
				this.host.dispatchEvent(evt);
			}
		}
		setHost(host, eventType) {
			this.host = host;
			if (eventType) this.eventType = eventType;
		}
	};
	var Context$2 = class {
		static {
			this.CHANGE_EVENT_TYPE = "un-context:change";
		}
		constructor(init, adoptedContext) {
			if (adoptedContext) {
				this.manager = adoptedContext.manager;
				this.object = adoptedContext.object;
				this.proxy = adoptedContext.proxy;
				this.update(init);
			} else {
				this.manager = new EffectsManager$2();
				this.object = init;
				this.proxy = createContext$2(this.object, this.manager);
			}
		}
		get(prop) {
			return this.proxy[prop];
		}
		set(prop, value) {
			this.proxy[prop] = value;
		}
		toObject() {
			return this.proxy;
		}
		update(next) {
			Object.assign(this.proxy, next);
		}
		apply(mapFn) {
			this.update(mapFn(this.proxy));
		}
		createEffect(fn) {
			createEffect$2(fn, this);
		}
		setHost(host, eventType) {
			this.manager.setHost(host, eventType);
		}
		open(effect) {
			this.manager.push(effect);
			return this.proxy;
		}
		close() {
			this.manager.pop();
		}
	};
	function createContext$2(root, manager) {
		return new Proxy(root, {
			get: (subject, prop, receiver) => {
				const value = Reflect.get(subject, prop, receiver);
				if (manager.isRunning() && isObservable$2(value)) manager.subscribe(prop, subject);
				return value;
			},
			set: (subject, prop, newValue, receiver) => {
				const didSet = Reflect.set(subject, prop, newValue, receiver);
				if (didSet && isObservable$2(newValue)) manager.runEffects(prop, subject);
				return didSet;
			}
		});
	}
	function isObservable$2(value) {
		switch (typeof value) {
			case "object":
			case "number":
			case "string":
			case "symbol":
			case "boolean":
			case "undefined": return true;
			default: return false;
		}
	}
	function createTemplate$2(fragment, render) {
		Object.assign(fragment, { render });
		return fragment;
	}
	function renderForEffects$2(original, effectors, ...scope) {
		const fragment = original.cloneNode(true);
		Array.from(effectors.entries()).forEach(([label, mutations]) => {
			const site = fragment.querySelector(`[data-${label}]`);
			if (site) mutations.forEach((fn) => fn(site, fragment, ...scope));
		});
		return fragment;
	}
	var TemplateParser$2 = class TemplateParser {
		static {
			this.parser = new DOMParser();
		}
		constructor(docType) {
			this.docType = "text/html";
			this.plugins = [];
			if (docType) this.docType = docType;
		}
		use(plugin) {
			this.plugins = this.plugins.concat(plugin);
		}
		parse(template, params) {
			const postProcess = {};
			const stringToParse = template.map((s, i) => {
				if (i >= params.length) return [s];
				const param = params[i];
				const place = this.classifyPlace(i, template);
				const mutation = this.tryReplacements(place, param);
				if (mutation) {
					const post = postProcess[place.nodeLabel];
					if (post) post.push(mutation);
					else postProcess[place.nodeLabel] = [mutation];
					switch (place.kind) {
						case "attr value": return [s, `"" data-${place.nodeLabel}`];
						case "tag content": return [s, `data-${place.nodeLabel}`];
						case "element content": return [s, `<ins data-${place.nodeLabel}></ins>`];
					}
				} else {
					console.error("No match for template parameter", place, param);
					throw `Failed to render template parameter ${i} around ${s}`;
				}
				return [s];
			}).flat().join("");
			const doc = TemplateParser.parser.parseFromString(stringToParse, this.docType);
			const collection = doc.head.childElementCount ? doc.head.children : doc.body.children;
			const fragment = new DocumentFragment();
			fragment.replaceChildren(...collection);
			const effectors = /* @__PURE__ */ new Map();
			for (const label in postProcess) {
				const site = fragment.querySelector(`[data-${label}]`);
				if (site) postProcess[label].forEach((m) => {
					const effector = m.apply(site, fragment);
					if (effector) {
						let list = effectors.get(label);
						if (list) list.push(effector);
						else effectors.set(label, [effector]);
					}
				});
			}
			return createTemplate$2(fragment, (...scope) => renderForEffects$2(fragment, effectors, ...scope));
		}
		static {
			this.OPEN_RE = /<([a-zA-z][$a-zA-Z0-9.-]*)\s+[^>]*$/;
		}
		static {
			this.IN_TAG_RE = /^(\s+|[^<>]*|"[^"]*")*$/;
		}
		static {
			this.ATTR_RE = /([$.]?[a-zA-Z][$a-zA-Z0-9.-]*)=\s*$/;
		}
		static {
			this.CLOSE_RE = /[/]?>[^<]*$/;
		}
		classifyPlace(i, template) {
			let tagOpen = null;
			for (let prev = i; prev >= 0; prev--) {
				if (template[prev].match(TemplateParser.CLOSE_RE)) break;
				tagOpen = template[prev].match(TemplateParser.OPEN_RE);
				if (tagOpen) break;
				if (!template[prev].match(TemplateParser.IN_TAG_RE)) break;
			}
			if (tagOpen) {
				const tagAttr = template[i].match(TemplateParser.ATTR_RE);
				if (tagAttr) return {
					kind: "attr value",
					nodeLabel: `node${i}`,
					tagName: tagOpen[1],
					attrName: tagAttr[1]
				};
				return {
					kind: "tag content",
					nodeLabel: `node${i}`,
					tagName: tagOpen[1]
				};
			}
			return {
				kind: "element content",
				nodeLabel: `node${i}`
			};
		}
		tryReplacements(place, param) {
			const replacements = this.plugins;
			for (let i = 0; i < replacements.length; i++) {
				const sub = replacements[i];
				if (place.kind === sub.place && checkType$2(param, sub)) return sub.mutator(place, param);
			}
		}
	};
	function checkType$2(param, sub) {
		if (typeof sub.types === "function") return sub.types(param, sub);
		return sub.types.includes(typeof param);
	}
	var Mutation$2 = class {
		constructor(place) {
			this.place = place;
		}
		apply(_site, _fragment) {
			throw "abstract method 'apply' called";
		}
	};
	var ElementContentMutation$2 = class extends Mutation$2 {
		constructor(place, content) {
			super(place);
			this.content = content;
		}
		apply(site, fragment) {
			(site.parentNode || fragment).replaceChild(this.content, site);
		}
	};
	var AttributeValueMutation$2 = class extends Mutation$2 {
		constructor(place, text) {
			super(place);
			this.text = text;
			this.name = place.attrName;
		}
		apply(site) {
			site.setAttribute(this.name, this.text);
		}
	};
	var ElementContentEffect$2 = class extends Mutation$2 {
		constructor(place, fn) {
			super(place);
			this.fn = fn;
		}
		apply(_site, _fragment) {
			const key = this.place.nodeLabel;
			return (site, fragment, ...scope) => {
				const start = new Comment(` <<< ${key} `);
				const end = new Comment(` >>> ${key} `);
				const placeholder = new DocumentFragment();
				placeholder.replaceChildren(start, end);
				(site.parentNode || fragment).replaceChild(placeholder, site);
				createEffect$2((...args) => {
					replaceElementContent$2(this.fn(...args), start, end);
				}, ...scope);
			};
		}
	};
	function replaceElementContent$2(value, start, end) {
		const parent = start.parentNode;
		if (!parent) throw new Error("No parent for placeholder");
		const valueToNode = (v) => {
			if (Array.isArray(v)) {
				const frag = new DocumentFragment();
				const nodes = v.map(valueToNode);
				frag.replaceChildren(...nodes);
				return frag;
			} else if (v instanceof Node) return v;
			else return new Text(v?.toString() || "");
		};
		const node = valueToNode(value);
		let p = start.nextSibling;
		while (p && p !== end) {
			const old = p;
			p = p.nextSibling;
			parent.removeChild(old);
		}
		if (node) parent.insertBefore(node, end);
	}
	var AttributeValueEffect$2 = class extends Mutation$2 {
		constructor(place, fn) {
			super(place);
			this.fn = fn;
			this.name = place.attrName;
		}
		apply(_site, _fragment) {
			return (site, _, ...scope) => createEffect$2((...args) => {
				replaceAttributeValue$2(this.fn(...args), site, this.name);
			}, ...scope);
		}
	};
	function replaceAttributeValue$2(value, site, attrName) {
		const special = attrName.match(/^([.$])(.+)$/);
		if (special) {
			const [_, pre, name] = special;
			switch (pre) {
				case ".":
					site[name] = value;
					break;
				case "$":
					if ("viewModel" in site && site.viewModel instanceof Context$2) site.viewModel.set(name, value);
					break;
			}
		} else switch (typeof value) {
			case "string":
				site.setAttribute(attrName, value);
				break;
			case "undefined":
			case "object":
			case "boolean":
				if (value) site.setAttribute(attrName, attrName);
				else site.removeAttribute(attrName);
				break;
			default: site.setAttribute(attrName, value?.toString());
		}
	}
	var TagReferenceEffect$2 = class extends Mutation$2 {
		constructor(place, fn) {
			super(place);
			this.fn = fn;
		}
		apply(_site, _fragment) {
			return (site, _, ...scope) => createEffect$2((...args) => {
				const value = this.fn(...args);
				if (typeof value === "function") value(site);
			}, ...scope);
		}
	};
	var htmlParser = new TemplateParser$2();
	htmlParser.use([
		{
			place: "element content",
			types: [
				"string",
				"number",
				"bigint",
				"symbol",
				"boolean"
			],
			mutator: (place, value) => new ElementContentMutation$2(place, new Text(value?.toString() || ""))
		},
		{
			place: "attr value",
			types: [
				"string",
				"number",
				"bigint",
				"symbol"
			],
			mutator: (place, value) => new AttributeValueMutation$2(place, value?.toString() || "")
		},
		{
			place: "element content",
			types: (param) => param instanceof Node,
			mutator: (place, value) => new ElementContentMutation$2(place, value)
		},
		{
			place: "element content",
			types: (param) => Array.isArray(param),
			mutator: (place, value) => {
				const fragment = new DocumentFragment();
				const nodes = value.map((v) => v instanceof Node ? v : new Text(v?.toString() || ""));
				fragment.append(...nodes);
				return new ElementContentMutation$2(place, fragment);
			}
		},
		{
			place: "element content",
			types: ["function"],
			mutator: (place, param) => new ElementContentEffect$2(place, param)
		},
		{
			place: "attr value",
			types: ["function"],
			mutator: (place, param) => new AttributeValueEffect$2(place, param)
		},
		{
			place: "tag content",
			types: ["function"],
			mutator: (place, param) => new TagReferenceEffect$2(place, param)
		}
	]);
	function html(template, ...params) {
		return htmlParser.parse(template, params);
	}
	/**
	* Memory leak warning!
	* Need to also clean up all the event listeners, probably
	* on disconnectCallback();
	*/
	function listen(element, map) {
		for (const eventType in map) element.addEventListener(eventType, map[eventType]);
	}
	function delegate(element, selector, map) {
		for (const eventType in map) {
			const listener = function(ev) {
				const target = ev.target;
				if (target && target instanceof HTMLElement && (target.matches(selector) || element.contains(target.closest(selector)))) map[eventType](ev);
			};
			element.addEventListener(eventType, listener);
		}
	}
	function relay(event, customType, detail) {
		const relay = event.target;
		const customEvent = new CustomEvent(customType, {
			bubbles: true,
			composed: true,
			detail
		});
		relay.dispatchEvent(customEvent);
		event.stopPropagation();
	}
	var Events = {
		listen,
		delegate,
		relay
	};
	function shadow(el, options = { mode: "open" }) {
		const shadowRoot = el.shadowRoot || el.attachShadow(options);
		const chain = {
			template,
			styles,
			replace,
			root: shadowRoot,
			delegate,
			listen
		};
		return chain;
		function template(fragment) {
			const first = fragment.firstElementChild;
			const template = first && first.tagName === "TEMPLATE" ? first : void 0;
			if (template) shadowRoot.appendChild(template.content.cloneNode(true));
			return chain;
		}
		function styles(...sheets) {
			shadowRoot.adoptedStyleSheets = sheets;
			return chain;
		}
		function replace(fragment) {
			shadowRoot.replaceChildren(fragment);
			return chain;
		}
		function listen(map) {
			Events.listen(shadowRoot, map);
			return chain;
		}
		function delegate(selector, map) {
			Events.delegate(shadowRoot, selector, map);
			return chain;
		}
	}
	//#endregion
	//#region ../view/dist/view.js
	var MappedSource = class {
		constructor(source, mapping) {
			this.initialized = false;
			this.origin = source;
			const entries = mapEntries(mapping).map(([t, s]) => [s, t]);
			this.inverse = Object.fromEntries(entries);
		}
		mapObservation(source) {
			const entries = Object.entries(source).map(([s, v]) => {
				const t = this.inverse[s];
				return t ? [t, v] : [];
			});
			return Object.fromEntries(entries.filter((pair) => pair.length > 0));
		}
		start(fn) {
			const sfn = (s, value) => {
				const t = this.inverse[s];
				fn(t, value);
				this.initialized = true;
			};
			return this.origin.start(sfn).then((obs) => this.initialized ? {} : this.mapObservation(obs));
		}
	};
	function mapEntries(mapping) {
		return Object.entries(mapping).map(([k, v]) => [k, v]);
	}
	var SignalEvent$1 = class extends CustomEvent {
		constructor(eventType, signal) {
			super(eventType, {
				bubbles: true,
				cancelable: true,
				composed: true,
				detail: signal
			});
		}
	};
	function createEffect$1(fn, ...scope) {
		const effect = { execute() {
			console.log("▶️ Executing created effect:", scope, fn);
			fn(...scope.map((cx) => cx.open(effect)));
			scope.forEach((cx) => cx.close());
		} };
		effect.execute();
	}
	var Scheduler$1 = class Scheduler {
		constructor() {
			this.signals = /* @__PURE__ */ new WeakMap();
			this.scheduled = /* @__PURE__ */ new WeakSet();
		}
		static {
			this.scheduler = new Scheduler();
		}
		subscribe(scope, key, effect) {
			let signals = this.signals.get(scope);
			if (!signals) {
				signals = /* @__PURE__ */ new Map();
				this.signals.set(scope, signals);
			}
			let signal = signals.get(key);
			if (!signal) {
				signal = /* @__PURE__ */ new Set();
				signals.set(key, signal);
			}
			signal.add(effect);
		}
		scheduleEffects(scope, key) {
			const signals = this.signals.get(scope);
			if (!signals) return;
			const signal = signals.get(key);
			if (signal) {
				for (const effect of signal) if (!this.scheduled.has(effect)) {
					this.scheduled.add(effect);
					setTimeout(() => {
						this.scheduled.delete(effect);
						effect.execute();
					});
				}
			}
		}
	};
	var EffectsManager$1 = class {
		constructor() {
			this.running = [];
			this.eventType = "un-effect:change";
		}
		isRunning() {
			return this.running.length > 0;
		}
		push(effect) {
			this.running.push(effect);
		}
		pop() {
			this.running.pop();
		}
		current() {
			const len = this.running.length;
			return len ? this.running[len - 1] : void 0;
		}
		subscribe(key, scope) {
			const current = this.current();
			if (current) Scheduler$1.scheduler.subscribe(scope, key, current);
		}
		runEffects(key, scope) {
			console.log("⚙️ Running effects for signal", key, scope);
			Scheduler$1.scheduler.scheduleEffects(scope, key);
			if (this.host) {
				const evt = new SignalEvent$1(this.eventType, {
					property: key,
					value: scope[key]
				});
				this.host.dispatchEvent(evt);
			}
		}
		setHost(host, eventType) {
			this.host = host;
			if (eventType) this.eventType = eventType;
		}
	};
	var Context$1 = class {
		static {
			this.CHANGE_EVENT_TYPE = "un-context:change";
		}
		constructor(init, adoptedContext) {
			if (adoptedContext) {
				this.manager = adoptedContext.manager;
				this.object = adoptedContext.object;
				this.proxy = adoptedContext.proxy;
				this.update(init);
			} else {
				this.manager = new EffectsManager$1();
				this.object = init;
				this.proxy = createContext$1(this.object, this.manager);
			}
		}
		get(prop) {
			return this.proxy[prop];
		}
		set(prop, value) {
			this.proxy[prop] = value;
		}
		toObject() {
			return this.proxy;
		}
		update(next) {
			Object.assign(this.proxy, next);
		}
		apply(mapFn) {
			this.update(mapFn(this.proxy));
		}
		createEffect(fn) {
			createEffect$1(fn, this);
		}
		setHost(host, eventType) {
			this.manager.setHost(host, eventType);
		}
		open(effect) {
			this.manager.push(effect);
			return this.proxy;
		}
		close() {
			this.manager.pop();
		}
	};
	function createContext$1(root, manager) {
		return new Proxy(root, {
			get: (subject, prop, receiver) => {
				const value = Reflect.get(subject, prop, receiver);
				if (manager.isRunning() && isObservable$1(value)) manager.subscribe(prop, subject);
				return value;
			},
			set: (subject, prop, newValue, receiver) => {
				const didSet = Reflect.set(subject, prop, newValue, receiver);
				if (didSet && isObservable$1(newValue)) manager.runEffects(prop, subject);
				return didSet;
			}
		});
	}
	function isObservable$1(value) {
		switch (typeof value) {
			case "object":
			case "number":
			case "string":
			case "symbol":
			case "boolean":
			case "undefined": return true;
			default: return false;
		}
	}
	function createTemplate$1(fragment, render) {
		Object.assign(fragment, { render });
		return fragment;
	}
	function renderForEffects$1(original, effectors, ...scope) {
		const fragment = original.cloneNode(true);
		Array.from(effectors.entries()).forEach(([label, mutations]) => {
			const site = fragment.querySelector(`[data-${label}]`);
			if (site) mutations.forEach((fn) => fn(site, fragment, ...scope));
		});
		return fragment;
	}
	var TemplateParser$1 = class TemplateParser {
		static {
			this.parser = new DOMParser();
		}
		constructor(docType) {
			this.docType = "text/html";
			this.plugins = [];
			if (docType) this.docType = docType;
		}
		use(plugin) {
			this.plugins = this.plugins.concat(plugin);
		}
		parse(template, params) {
			const postProcess = {};
			const stringToParse = template.map((s, i) => {
				if (i >= params.length) return [s];
				const param = params[i];
				const place = this.classifyPlace(i, template);
				const mutation = this.tryReplacements(place, param);
				if (mutation) {
					const post = postProcess[place.nodeLabel];
					if (post) post.push(mutation);
					else postProcess[place.nodeLabel] = [mutation];
					switch (place.kind) {
						case "attr value": return [s, `"" data-${place.nodeLabel}`];
						case "tag content": return [s, `data-${place.nodeLabel}`];
						case "element content": return [s, `<ins data-${place.nodeLabel}></ins>`];
					}
				} else {
					console.error("No match for template parameter", place, param);
					throw `Failed to render template parameter ${i} around ${s}`;
				}
				return [s];
			}).flat().join("");
			const doc = TemplateParser.parser.parseFromString(stringToParse, this.docType);
			const collection = doc.head.childElementCount ? doc.head.children : doc.body.children;
			const fragment = new DocumentFragment();
			fragment.replaceChildren(...collection);
			const effectors = /* @__PURE__ */ new Map();
			for (const label in postProcess) {
				const site = fragment.querySelector(`[data-${label}]`);
				if (site) postProcess[label].forEach((m) => {
					const effector = m.apply(site, fragment);
					if (effector) {
						let list = effectors.get(label);
						if (list) list.push(effector);
						else effectors.set(label, [effector]);
					}
				});
			}
			return createTemplate$1(fragment, (...scope) => renderForEffects$1(fragment, effectors, ...scope));
		}
		static {
			this.OPEN_RE = /<([a-zA-z][$a-zA-Z0-9.-]*)\s+[^>]*$/;
		}
		static {
			this.IN_TAG_RE = /^(\s+|[^<>]*|"[^"]*")*$/;
		}
		static {
			this.ATTR_RE = /([$.]?[a-zA-Z][$a-zA-Z0-9.-]*)=\s*$/;
		}
		static {
			this.CLOSE_RE = /[/]?>[^<]*$/;
		}
		classifyPlace(i, template) {
			let tagOpen = null;
			for (let prev = i; prev >= 0; prev--) {
				if (template[prev].match(TemplateParser.CLOSE_RE)) break;
				tagOpen = template[prev].match(TemplateParser.OPEN_RE);
				if (tagOpen) break;
				if (!template[prev].match(TemplateParser.IN_TAG_RE)) break;
			}
			if (tagOpen) {
				const tagAttr = template[i].match(TemplateParser.ATTR_RE);
				if (tagAttr) return {
					kind: "attr value",
					nodeLabel: `node${i}`,
					tagName: tagOpen[1],
					attrName: tagAttr[1]
				};
				return {
					kind: "tag content",
					nodeLabel: `node${i}`,
					tagName: tagOpen[1]
				};
			}
			return {
				kind: "element content",
				nodeLabel: `node${i}`
			};
		}
		tryReplacements(place, param) {
			const replacements = this.plugins;
			for (let i = 0; i < replacements.length; i++) {
				const sub = replacements[i];
				if (place.kind === sub.place && checkType$1(param, sub)) return sub.mutator(place, param);
			}
		}
	};
	function checkType$1(param, sub) {
		if (typeof sub.types === "function") return sub.types(param, sub);
		return sub.types.includes(typeof param);
	}
	var Mutation$1 = class {
		constructor(place) {
			this.place = place;
		}
		apply(_site, _fragment) {
			throw "abstract method 'apply' called";
		}
	};
	var ElementContentMutation$1 = class extends Mutation$1 {
		constructor(place, content) {
			super(place);
			this.content = content;
		}
		apply(site, fragment) {
			(site.parentNode || fragment).replaceChild(this.content, site);
		}
	};
	var AttributeValueMutation$1 = class extends Mutation$1 {
		constructor(place, text) {
			super(place);
			this.text = text;
			this.name = place.attrName;
		}
		apply(site) {
			site.setAttribute(this.name, this.text);
		}
	};
	var ElementContentEffect$1 = class extends Mutation$1 {
		constructor(place, fn) {
			super(place);
			this.fn = fn;
		}
		apply(_site, _fragment) {
			const key = this.place.nodeLabel;
			return (site, fragment, ...scope) => {
				const start = new Comment(` <<< ${key} `);
				const end = new Comment(` >>> ${key} `);
				const placeholder = new DocumentFragment();
				placeholder.replaceChildren(start, end);
				(site.parentNode || fragment).replaceChild(placeholder, site);
				createEffect$1((...args) => {
					replaceElementContent$1(this.fn(...args), start, end);
				}, ...scope);
			};
		}
	};
	function replaceElementContent$1(value, start, end) {
		const parent = start.parentNode;
		if (!parent) throw new Error("No parent for placeholder");
		const valueToNode = (v) => {
			if (Array.isArray(v)) {
				const frag = new DocumentFragment();
				const nodes = v.map(valueToNode);
				frag.replaceChildren(...nodes);
				return frag;
			} else if (v instanceof Node) return v;
			else return new Text(v?.toString() || "");
		};
		const node = valueToNode(value);
		let p = start.nextSibling;
		while (p && p !== end) {
			const old = p;
			p = p.nextSibling;
			parent.removeChild(old);
		}
		if (node) parent.insertBefore(node, end);
	}
	var AttributeValueEffect$1 = class extends Mutation$1 {
		constructor(place, fn) {
			super(place);
			this.fn = fn;
			this.name = place.attrName;
		}
		apply(_site, _fragment) {
			return (site, _, ...scope) => createEffect$1((...args) => {
				replaceAttributeValue$1(this.fn(...args), site, this.name);
			}, ...scope);
		}
	};
	function replaceAttributeValue$1(value, site, attrName) {
		const special = attrName.match(/^([.$])(.+)$/);
		if (special) {
			const [_, pre, name] = special;
			switch (pre) {
				case ".":
					site[name] = value;
					break;
				case "$":
					if ("viewModel" in site && site.viewModel instanceof Context$1) site.viewModel.set(name, value);
					break;
			}
		} else switch (typeof value) {
			case "string":
				site.setAttribute(attrName, value);
				break;
			case "undefined":
			case "object":
			case "boolean":
				if (value) site.setAttribute(attrName, attrName);
				else site.removeAttribute(attrName);
				break;
			default: site.setAttribute(attrName, value?.toString());
		}
	}
	var TagReferenceEffect$1 = class extends Mutation$1 {
		constructor(place, fn) {
			super(place);
			this.fn = fn;
		}
		apply(_site, _fragment) {
			return (site, _, ...scope) => createEffect$1((...args) => {
				const value = this.fn(...args);
				if (typeof value === "function") value(site);
			}, ...scope);
		}
	};
	new TemplateParser$1().use([
		{
			place: "element content",
			types: [
				"string",
				"number",
				"bigint",
				"symbol",
				"boolean"
			],
			mutator: (place, value) => new ElementContentMutation$1(place, new Text(value?.toString() || ""))
		},
		{
			place: "attr value",
			types: [
				"string",
				"number",
				"bigint",
				"symbol"
			],
			mutator: (place, value) => new AttributeValueMutation$1(place, value?.toString() || "")
		},
		{
			place: "element content",
			types: (param) => param instanceof Node,
			mutator: (place, value) => new ElementContentMutation$1(place, value)
		},
		{
			place: "element content",
			types: (param) => Array.isArray(param),
			mutator: (place, value) => {
				const fragment = new DocumentFragment();
				const nodes = value.map((v) => v instanceof Node ? v : new Text(v?.toString() || ""));
				fragment.append(...nodes);
				return new ElementContentMutation$1(place, fragment);
			}
		},
		{
			place: "element content",
			types: ["function"],
			mutator: (place, param) => new ElementContentEffect$1(place, param)
		},
		{
			place: "attr value",
			types: ["function"],
			mutator: (place, param) => new AttributeValueEffect$1(place, param)
		},
		{
			place: "tag content",
			types: ["function"],
			mutator: (place, param) => new TagReferenceEffect$1(place, param)
		}
	]);
	var ViewModel = class extends Context$1 {
		constructor(init, adoptedContext) {
			super(init, adoptedContext);
		}
		get $() {
			return this.toObject();
		}
		with(source, ...keys) {
			const mapping = Object.fromEntries(keys.map((k) => [k, k]));
			return this.merge(new MappedSource(source, mapping));
		}
		withCalculated(source, mapping) {
			return this.merge(new MappedSource(source, mapping));
		}
		withRenamed(source, renaming) {
			return this.merge(new MappedSource(source, renaming));
		}
		merge(source) {
			if (source) {
				const initialized = /* @__PURE__ */ new Map();
				source.start((name, value) => {
					this.set(name, value);
					initialized.set(name, true);
				}).then((firstObservation) => {
					Object.keys(firstObservation).filter((k) => !initialized.get(k)).forEach((k) => {
						this.set(k, firstObservation[k]);
					});
				});
			}
			return this;
		}
		render(template) {
			return template.render(this);
		}
	};
	function createViewModel(init) {
		if (init === void 0) return new ViewModel({});
		else return new ViewModel(init);
	}
	//#endregion
	//#region ../auth/dist/auth.js
	var __defProp$1 = Object.defineProperty;
	var __exportAll$1 = (all, no_symbols) => {
		let target = {};
		for (var name in all) __defProp$1(target, name, {
			get: all[name],
			enumerable: true
		});
		if (!no_symbols) __defProp$1(target, Symbol.toStringTag, { value: "Module" });
		return target;
	};
	var InvalidTokenError = class extends Error {};
	InvalidTokenError.prototype.name = "InvalidTokenError";
	function b64DecodeUnicode(str) {
		return decodeURIComponent(atob(str).replace(/(.)/g, (m, p) => {
			let code = p.charCodeAt(0).toString(16).toUpperCase();
			if (code.length < 2) code = "0" + code;
			return "%" + code;
		}));
	}
	function base64UrlDecode(str) {
		let output = str.replace(/-/g, "+").replace(/_/g, "/");
		switch (output.length % 4) {
			case 0: break;
			case 2:
				output += "==";
				break;
			case 3:
				output += "=";
				break;
			default: throw new Error("base64 string is not of the correct length");
		}
		try {
			return b64DecodeUnicode(output);
		} catch (err) {
			return atob(output);
		}
	}
	function jwtDecode(token, options) {
		if (typeof token !== "string") throw new InvalidTokenError("Invalid token specified: must be a string");
		options || (options = {});
		const pos = options.header === true ? 0 : 1;
		const part = token.split(".")[pos];
		if (typeof part !== "string") throw new InvalidTokenError(`Invalid token specified: missing part #${pos + 1}`);
		let decoded;
		try {
			decoded = base64UrlDecode(part);
		} catch (e) {
			throw new InvalidTokenError(`Invalid token specified: invalid base64 for part #${pos + 1} (${e.message})`);
		}
		try {
			return JSON.parse(decoded);
		} catch (e) {
			throw new InvalidTokenError(`Invalid token specified: invalid json for part #${pos + 1} (${e.message})`);
		}
	}
	var __defProp = Object.defineProperty;
	var __exportAll = (all, no_symbols) => {
		let target = {};
		for (var name in all) __defProp(target, name, {
			get: all[name],
			enumerable: true
		});
		if (!no_symbols) __defProp(target, Symbol.toStringTag, { value: "Module" });
		return target;
	};
	var message_exports = /* @__PURE__ */ __exportAll({
		Dispatch: () => Dispatch,
		None: () => None,
		dispatch: () => dispatch$1,
		dispatcher: () => dispatcher
	});
	var None = [];
	var Dispatch = class extends CustomEvent {
		constructor(msg, eventType = "un:message") {
			super(eventType, {
				bubbles: true,
				composed: true,
				detail: msg
			});
		}
	};
	function dispatcher(eventType = "un:message") {
		return (target, ...msg) => target.dispatchEvent(new Dispatch(msg, eventType));
	}
	var dispatch$1 = dispatcher();
	var SignalEvent = class extends CustomEvent {
		constructor(eventType, signal) {
			super(eventType, {
				bubbles: true,
				cancelable: true,
				composed: true,
				detail: signal
			});
		}
	};
	function createEffect(fn, ...scope) {
		const effect = { execute() {
			console.log("▶️ Executing created effect:", scope, fn);
			fn(...scope.map((cx) => cx.open(effect)));
			scope.forEach((cx) => cx.close());
		} };
		effect.execute();
	}
	var Scheduler = class Scheduler {
		constructor() {
			this.signals = /* @__PURE__ */ new WeakMap();
			this.scheduled = /* @__PURE__ */ new WeakSet();
		}
		static {
			this.scheduler = new Scheduler();
		}
		subscribe(scope, key, effect) {
			let signals = this.signals.get(scope);
			if (!signals) {
				signals = /* @__PURE__ */ new Map();
				this.signals.set(scope, signals);
			}
			let signal = signals.get(key);
			if (!signal) {
				signal = /* @__PURE__ */ new Set();
				signals.set(key, signal);
			}
			signal.add(effect);
		}
		scheduleEffects(scope, key) {
			const signals = this.signals.get(scope);
			if (!signals) return;
			const signal = signals.get(key);
			if (signal) {
				for (const effect of signal) if (!this.scheduled.has(effect)) {
					this.scheduled.add(effect);
					setTimeout(() => {
						this.scheduled.delete(effect);
						effect.execute();
					});
				}
			}
		}
	};
	var EffectsManager = class {
		constructor() {
			this.running = [];
			this.eventType = "un-effect:change";
		}
		isRunning() {
			return this.running.length > 0;
		}
		push(effect) {
			this.running.push(effect);
		}
		pop() {
			this.running.pop();
		}
		current() {
			const len = this.running.length;
			return len ? this.running[len - 1] : void 0;
		}
		subscribe(key, scope) {
			const current = this.current();
			if (current) Scheduler.scheduler.subscribe(scope, key, current);
		}
		runEffects(key, scope) {
			console.log("⚙️ Running effects for signal", key, scope);
			Scheduler.scheduler.scheduleEffects(scope, key);
			if (this.host) {
				const evt = new SignalEvent(this.eventType, {
					property: key,
					value: scope[key]
				});
				this.host.dispatchEvent(evt);
			}
		}
		setHost(host, eventType) {
			this.host = host;
			if (eventType) this.eventType = eventType;
		}
	};
	var Context = class {
		static {
			this.CHANGE_EVENT_TYPE = "un-context:change";
		}
		constructor(init, adoptedContext) {
			if (adoptedContext) {
				this.manager = adoptedContext.manager;
				this.object = adoptedContext.object;
				this.proxy = adoptedContext.proxy;
				this.update(init);
			} else {
				this.manager = new EffectsManager();
				this.object = init;
				this.proxy = createContext(this.object, this.manager);
			}
		}
		get(prop) {
			return this.proxy[prop];
		}
		set(prop, value) {
			this.proxy[prop] = value;
		}
		toObject() {
			return this.proxy;
		}
		update(next) {
			Object.assign(this.proxy, next);
		}
		apply(mapFn) {
			this.update(mapFn(this.proxy));
		}
		createEffect(fn) {
			createEffect(fn, this);
		}
		setHost(host, eventType) {
			this.manager.setHost(host, eventType);
		}
		open(effect) {
			this.manager.push(effect);
			return this.proxy;
		}
		close() {
			this.manager.pop();
		}
	};
	function createContext(root, manager) {
		return new Proxy(root, {
			get: (subject, prop, receiver) => {
				const value = Reflect.get(subject, prop, receiver);
				if (manager.isRunning() && isObservable(value)) manager.subscribe(prop, subject);
				return value;
			},
			set: (subject, prop, newValue, receiver) => {
				const didSet = Reflect.set(subject, prop, newValue, receiver);
				if (didSet && isObservable(newValue)) manager.runEffects(prop, subject);
				return didSet;
			}
		});
	}
	function isObservable(value) {
		switch (typeof value) {
			case "object":
			case "number":
			case "string":
			case "symbol":
			case "boolean":
			case "undefined": return true;
			default: return false;
		}
	}
	var DirectEffect = class {
		constructor(fn, ...scope) {
			this.effectFn = () => fn(...scope);
		}
		execute() {
			this.effectFn();
		}
	};
	function createTemplate(fragment, render) {
		Object.assign(fragment, { render });
		return fragment;
	}
	function renderForEffects(original, effectors, ...scope) {
		const fragment = original.cloneNode(true);
		Array.from(effectors.entries()).forEach(([label, mutations]) => {
			const site = fragment.querySelector(`[data-${label}]`);
			if (site) mutations.forEach((fn) => fn(site, fragment, ...scope));
		});
		return fragment;
	}
	var TemplateParser = class TemplateParser {
		static {
			this.parser = new DOMParser();
		}
		constructor(docType) {
			this.docType = "text/html";
			this.plugins = [];
			if (docType) this.docType = docType;
		}
		use(plugin) {
			this.plugins = this.plugins.concat(plugin);
		}
		parse(template, params) {
			const postProcess = {};
			const stringToParse = template.map((s, i) => {
				if (i >= params.length) return [s];
				const param = params[i];
				const place = this.classifyPlace(i, template);
				const mutation = this.tryReplacements(place, param);
				if (mutation) {
					const post = postProcess[place.nodeLabel];
					if (post) post.push(mutation);
					else postProcess[place.nodeLabel] = [mutation];
					switch (place.kind) {
						case "attr value": return [s, `"" data-${place.nodeLabel}`];
						case "tag content": return [s, `data-${place.nodeLabel}`];
						case "element content": return [s, `<ins data-${place.nodeLabel}></ins>`];
					}
				} else {
					console.error("No match for template parameter", place, param);
					throw `Failed to render template parameter ${i} around ${s}`;
				}
				return [s];
			}).flat().join("");
			const doc = TemplateParser.parser.parseFromString(stringToParse, this.docType);
			const collection = doc.head.childElementCount ? doc.head.children : doc.body.children;
			const fragment = new DocumentFragment();
			fragment.replaceChildren(...collection);
			const effectors = /* @__PURE__ */ new Map();
			for (const label in postProcess) {
				const site = fragment.querySelector(`[data-${label}]`);
				if (site) postProcess[label].forEach((m) => {
					const effector = m.apply(site, fragment);
					if (effector) {
						let list = effectors.get(label);
						if (list) list.push(effector);
						else effectors.set(label, [effector]);
					}
				});
			}
			return createTemplate(fragment, (...scope) => renderForEffects(fragment, effectors, ...scope));
		}
		static {
			this.OPEN_RE = /<([a-zA-z][$a-zA-Z0-9.-]*)\s+[^>]*$/;
		}
		static {
			this.IN_TAG_RE = /^(\s+|[^<>]*|"[^"]*")*$/;
		}
		static {
			this.ATTR_RE = /([$.]?[a-zA-Z][$a-zA-Z0-9.-]*)=\s*$/;
		}
		static {
			this.CLOSE_RE = /[/]?>[^<]*$/;
		}
		classifyPlace(i, template) {
			let tagOpen = null;
			for (let prev = i; prev >= 0; prev--) {
				if (template[prev].match(TemplateParser.CLOSE_RE)) break;
				tagOpen = template[prev].match(TemplateParser.OPEN_RE);
				if (tagOpen) break;
				if (!template[prev].match(TemplateParser.IN_TAG_RE)) break;
			}
			if (tagOpen) {
				const tagAttr = template[i].match(TemplateParser.ATTR_RE);
				if (tagAttr) return {
					kind: "attr value",
					nodeLabel: `node${i}`,
					tagName: tagOpen[1],
					attrName: tagAttr[1]
				};
				return {
					kind: "tag content",
					nodeLabel: `node${i}`,
					tagName: tagOpen[1]
				};
			}
			return {
				kind: "element content",
				nodeLabel: `node${i}`
			};
		}
		tryReplacements(place, param) {
			const replacements = this.plugins;
			for (let i = 0; i < replacements.length; i++) {
				const sub = replacements[i];
				if (place.kind === sub.place && checkType(param, sub)) return sub.mutator(place, param);
			}
		}
	};
	function checkType(param, sub) {
		if (typeof sub.types === "function") return sub.types(param, sub);
		return sub.types.includes(typeof param);
	}
	var Mutation = class {
		constructor(place) {
			this.place = place;
		}
		apply(_site, _fragment) {
			throw "abstract method 'apply' called";
		}
	};
	var ElementContentMutation = class extends Mutation {
		constructor(place, content) {
			super(place);
			this.content = content;
		}
		apply(site, fragment) {
			(site.parentNode || fragment).replaceChild(this.content, site);
		}
	};
	var AttributeValueMutation = class extends Mutation {
		constructor(place, text) {
			super(place);
			this.text = text;
			this.name = place.attrName;
		}
		apply(site) {
			site.setAttribute(this.name, this.text);
		}
	};
	var ElementContentEffect = class extends Mutation {
		constructor(place, fn) {
			super(place);
			this.fn = fn;
		}
		apply(_site, _fragment) {
			const key = this.place.nodeLabel;
			return (site, fragment, ...scope) => {
				const start = new Comment(` <<< ${key} `);
				const end = new Comment(` >>> ${key} `);
				const placeholder = new DocumentFragment();
				placeholder.replaceChildren(start, end);
				(site.parentNode || fragment).replaceChild(placeholder, site);
				createEffect((...args) => {
					replaceElementContent(this.fn(...args), start, end);
				}, ...scope);
			};
		}
	};
	function replaceElementContent(value, start, end) {
		const parent = start.parentNode;
		if (!parent) throw new Error("No parent for placeholder");
		const valueToNode = (v) => {
			if (Array.isArray(v)) {
				const frag = new DocumentFragment();
				const nodes = v.map(valueToNode);
				frag.replaceChildren(...nodes);
				return frag;
			} else if (v instanceof Node) return v;
			else return new Text(v?.toString() || "");
		};
		const node = valueToNode(value);
		let p = start.nextSibling;
		while (p && p !== end) {
			const old = p;
			p = p.nextSibling;
			parent.removeChild(old);
		}
		if (node) parent.insertBefore(node, end);
	}
	var AttributeValueEffect = class extends Mutation {
		constructor(place, fn) {
			super(place);
			this.fn = fn;
			this.name = place.attrName;
		}
		apply(_site, _fragment) {
			return (site, _, ...scope) => createEffect((...args) => {
				replaceAttributeValue(this.fn(...args), site, this.name);
			}, ...scope);
		}
	};
	function replaceAttributeValue(value, site, attrName) {
		const special = attrName.match(/^([.$])(.+)$/);
		if (special) {
			const [_, pre, name] = special;
			switch (pre) {
				case ".":
					site[name] = value;
					break;
				case "$":
					if ("viewModel" in site && site.viewModel instanceof Context) site.viewModel.set(name, value);
					break;
			}
		} else switch (typeof value) {
			case "string":
				site.setAttribute(attrName, value);
				break;
			case "undefined":
			case "object":
			case "boolean":
				if (value) site.setAttribute(attrName, attrName);
				else site.removeAttribute(attrName);
				break;
			default: site.setAttribute(attrName, value?.toString());
		}
	}
	var TagReferenceEffect = class extends Mutation {
		constructor(place, fn) {
			super(place);
			this.fn = fn;
		}
		apply(_site, _fragment) {
			return (site, _, ...scope) => createEffect((...args) => {
				const value = this.fn(...args);
				if (typeof value === "function") value(site);
			}, ...scope);
		}
	};
	new TemplateParser().use([
		{
			place: "element content",
			types: [
				"string",
				"number",
				"bigint",
				"symbol",
				"boolean"
			],
			mutator: (place, value) => new ElementContentMutation(place, new Text(value?.toString() || ""))
		},
		{
			place: "attr value",
			types: [
				"string",
				"number",
				"bigint",
				"symbol"
			],
			mutator: (place, value) => new AttributeValueMutation(place, value?.toString() || "")
		},
		{
			place: "element content",
			types: (param) => param instanceof Node,
			mutator: (place, value) => new ElementContentMutation(place, value)
		},
		{
			place: "element content",
			types: (param) => Array.isArray(param),
			mutator: (place, value) => {
				const fragment = new DocumentFragment();
				const nodes = value.map((v) => v instanceof Node ? v : new Text(v?.toString() || ""));
				fragment.append(...nodes);
				return new ElementContentMutation(place, fragment);
			}
		},
		{
			place: "element content",
			types: ["function"],
			mutator: (place, param) => new ElementContentEffect(place, param)
		},
		{
			place: "attr value",
			types: ["function"],
			mutator: (place, param) => new AttributeValueEffect(place, param)
		},
		{
			place: "tag content",
			types: ["function"],
			mutator: (place, param) => new TagReferenceEffect(place, param)
		}
	]);
	var Provider = class Provider extends HTMLElement {
		static {
			this.DISCOVERY_EVENT = "un-provider:discover";
		}
		static {
			this.REGISTRY_EVENT = "un-provider:register";
		}
		static {
			this.CHANGE_EVENT = "un-provider:change";
		}
		static {
			document.addEventListener(Provider.DISCOVERY_EVENT, (event) => {
				const [contextLabel, respondFn] = event.detail;
				const provider = registeredProvider(contextLabel);
				if (provider) respondFn(provider);
			});
			document.addEventListener(Provider.REGISTRY_EVENT, (event) => {
				const [contextLabel, provider] = event.detail;
				registerProvider(contextLabel, provider);
			});
		}
		constructor(init, label) {
			super();
			this.contextLabel = label;
			this.context = new Context(init);
			this.context.setHost(this, Provider.CHANGE_EVENT);
			this.addEventListener(Provider.DISCOVERY_EVENT, (event) => {
				const [contextLabel, respondFn] = event.detail;
				if (contextLabel === this.contextLabel) {
					event.stopPropagation();
					respondFn(this);
				}
			});
			const registryEvent = new CustomEvent(Provider.REGISTRY_EVENT, {
				bubbles: true,
				composed: true,
				detail: [this.contextLabel, this]
			});
			this.dispatchEvent(registryEvent);
		}
		attach(observer) {
			this.addEventListener(Provider.CHANGE_EVENT, observer);
			return this.context.toObject();
		}
		detach(observer) {
			this.removeEventListener(Provider.CHANGE_EVENT, observer);
		}
	};
	function discover(observer, contextLabel) {
		return new Promise((resolve, reject) => {
			const discoveryEvent = new CustomEvent(Provider.DISCOVERY_EVENT, {
				bubbles: true,
				composed: true,
				detail: [contextLabel, (provider) => provider ? resolve(provider) : reject()]
			});
			if (observer.isConnected) observer.dispatchEvent(discoveryEvent);
			else document.dispatchEvent(discoveryEvent);
		});
	}
	var registry = {};
	function registerProvider(label, provider) {
		registry[label] = provider;
	}
	function registeredProvider(label) {
		return registry[label];
	}
	var Observer = class {
		constructor(contextLabel) {
			this.contextLabel = contextLabel;
		}
		observe(from, fn) {
			return new Promise((resolve, reject) => {
				if (this.provider) resolve(this.attachObserver(fn));
				else discover(from, this.contextLabel).then((provider) => {
					this.provider = provider;
					resolve(this.attachObserver(fn));
				}).catch((err) => reject(err));
			});
		}
		attachObserver(fn) {
			const init = this.provider.attach((ev) => {
				const { property, value } = ev.detail;
				const effect = new DirectEffect(fn, {
					property,
					value
				});
				console.log("📡Signal received:", property, value);
				effect.execute();
			});
			console.log("👀 Initial observation:", init);
			return init;
		}
	};
	var Service = class {
		constructor(update, context, eventType = "service:message", autostart = true) {
			this._pending = [];
			this._context = context;
			this._update = update;
			this._eventType = eventType;
			this._running = autostart;
		}
		attach(host) {
			host.addEventListener(this._eventType, (ev) => {
				ev.stopPropagation();
				const message = ev.detail;
				this.consume(message);
			});
		}
		start() {
			if (!this._running) {
				console.log(`Starting ${this._eventType} service`);
				this._running = true;
				this._pending.forEach((msg) => this.process(msg));
			}
		}
		consume(message) {
			if (message.length !== 0) {
				const trueMsg = message;
				if (this._running) this.process(trueMsg);
				else {
					console.log(`📥 Queueing ${this._eventType} message`, message);
					this._pending.push(trueMsg);
				}
			}
		}
		process(message) {
			console.log(`📤 Processing ${this._eventType} message`, message);
			const next = this._update(message, this._context.toObject());
			console.log("⏭️ next state: ", next);
			if (!Array.isArray(next)) this._context.update(next);
			else {
				const [now, ...later] = next;
				later.forEach((promise) => promise.then((command) => this.consume(command)));
				this._context.update(now);
			}
		}
	};
	function fromService(target, contextLabel) {
		return new FromService(target, contextLabel);
	}
	var FromService = class {
		constructor(client, contextLabel) {
			this.client = client;
			this.observer = new Observer(contextLabel);
		}
		start(fn) {
			return this.observer.observe(this.client, (s) => {
				fn(s.property, s.value);
			});
		}
	};
	var auth_exports = /* @__PURE__ */ __exportAll$1({
		AuthenticatedUser: () => AuthenticatedUser,
		CONTEXT_DEFAULT: () => AUTH_CONTEXT_DEFAULT,
		Provider: () => AuthProvider,
		User: () => APIUser,
		dispatch: () => dispatch,
		headers: () => authHeaders,
		payload: () => tokenPayload
	});
	var AUTH_CONTEXT_DEFAULT = "context:auth";
	var APIUser = class APIUser {
		static {
			this.TOKEN_KEY = "un-auth:token";
		}
		constructor(username) {
			this.authenticated = false;
			this.username = username || "anonymous";
		}
		static deauthenticate(user) {
			user.authenticated = false;
			user.username = "anonymous";
			localStorage.removeItem(APIUser.TOKEN_KEY);
			return user;
		}
	};
	var AuthenticatedUser = class AuthenticatedUser extends APIUser {
		constructor(token) {
			super();
			const jsonPayload = jwtDecode(token);
			console.log("Token payload", jsonPayload);
			this.token = token;
			this.authenticated = true;
			this.username = jsonPayload.username;
		}
		static authenticate(token) {
			const authenticatedUser = new AuthenticatedUser(token);
			localStorage.setItem(APIUser.TOKEN_KEY, token);
			return authenticatedUser;
		}
		static authenticateFromLocalStorage() {
			const priorToken = localStorage.getItem(APIUser.TOKEN_KEY);
			return priorToken ? AuthenticatedUser.authenticate(priorToken) : new APIUser();
		}
	};
	var AuthService = class AuthService extends Service {
		static {
			this.EVENT_TYPE = "auth:message";
		}
		constructor(context, redirectForLogin) {
			super((msg, model) => this.update(msg, model), context, AuthService.EVENT_TYPE);
			this._redirectForLogin = redirectForLogin;
		}
		update(message, model) {
			console.log("Auth update", message);
			switch (message[0]) {
				case "auth/signin":
					const { token, redirect } = message[1];
					return [signIn(token), redirectLater(redirect)];
				case "auth/signout": return [signOut(model), redirectLater(this._redirectForLogin)];
				case "auth/redirect":
					redirectNow(this._redirectForLogin, { next: window.location.href });
					break;
				default:
					const unhandled = message[0];
					throw new Error(`Unhandled Auth message "${unhandled}"`);
			}
			return model;
		}
	};
	var AuthProvider = class extends Provider {
		get redirect() {
			return this.getAttribute("redirect") || void 0;
		}
		constructor() {
			const user = AuthenticatedUser.authenticateFromLocalStorage();
			const { authenticated, username } = user;
			super({
				authenticated,
				username,
				token: authenticated ? user.token : void 0
			}, AUTH_CONTEXT_DEFAULT);
		}
		connectedCallback() {
			new AuthService(this.context, this.redirect).attach(this);
		}
	};
	var dispatch = message_exports.dispatcher(AuthService.EVENT_TYPE);
	function redirectLater(redirect, query = {}) {
		return new Promise((resolve) => {
			redirectNow(redirect, query);
			resolve(message_exports.None);
		});
	}
	function redirectNow(redirect, query = {}) {
		if (redirect) {
			const base = window.location.href;
			const target = new URL(redirect, base);
			Object.entries(query).forEach(([k, v]) => target.searchParams.set(k, v));
			console.log("Auth redirecting to ", redirect);
			window.history.pushState({}, "", redirect);
			window.history.go();
		}
	}
	function signIn(token) {
		const { authenticated, username } = AuthenticatedUser.authenticate(token);
		return {
			authenticated,
			username,
			token
		};
	}
	function signOut(model) {
		const { authenticated, username } = APIUser.deauthenticate(new APIUser(model.username));
		return {
			username,
			authenticated,
			token: void 0
		};
	}
	function authHeaders(auth) {
		if (auth.authenticated) return { Authorization: `Bearer ${auth.token || "NO_TOKEN"}` };
		else return {};
	}
	function tokenPayload(user) {
		if (user.authenticated) return jwtDecode(user.token || "");
		else return {};
	}
	function fromAuth(target, contextLabel = AUTH_CONTEXT_DEFAULT) {
		return fromService(target, contextLabel);
	}
	//#endregion
	//#region src/fromHistory.ts
	function fromHistory(target, contextLabel = HISTORY_CONTEXT_DEFAULT) {
		return fromService$1(target, contextLabel);
	}
	//#endregion
	//#region src/switch.ts
	var switch_exports = /* @__PURE__ */ __exportAll$3({
		Element: () => Switch,
		Switch: () => Switch
	});
	var Switch = class extends HTMLElement {
		constructor(routes) {
			super();
			this.viewModel = createViewModel({ authenticated: false }).with(fromAuth(this), "authenticated", "username").with(fromHistory(this), "location");
			this._cases = [];
			this._routeView = html`
    <h1>Routing...</h1>
  `;
			this._routeViewModel = createViewModel({
				params: {},
				query: new URLSearchParams()
			});
			this._cases = routes.map((r) => ({
				...r,
				route: new import_route_parser.default(r.path)
			}));
			this.viewModel.createEffect(($) => {
				console.log("📍 Router got location:", $.location);
				if ($.location) {
					const nextView = this.routeToView($.location, $.authenticated, $.username);
					if (nextView !== this._routeView) {
						this._routeView = nextView;
						shadow(this).replace(this._routeViewModel.render(nextView));
					}
				}
			});
		}
		routeToView(location, authenticated = false, username) {
			const m = this.matchRoute(location);
			if (m) {
				if ("view" in m) if (m.auth && m.auth !== "public" && !authenticated) {
					auth_exports.dispatch(this, "auth/redirect");
					return html`
            <h1>Redirecting for Login</h1>
          `;
				} else {
					this._routeViewModel.update({
						params: m.params,
						query: m.query,
						user: {
							authenticated,
							username
						}
					});
					return m.view;
				}
				if ("redirect" in m) {
					const redirect = m.redirect;
					if (typeof redirect === "string") {
						this.redirect(redirect);
						return html`
            <h1>Redirecting to ${redirect}…</h1>
          `;
					}
				}
			}
			return html`
      <h1>Not Found</h1>
    `;
		}
		matchRoute(location) {
			const { search, pathname } = location;
			const query = new URLSearchParams(search);
			const path = pathname + search;
			for (const option of this._cases) {
				const params = option.route.match(path);
				if (params) return {
					...option,
					path: pathname,
					params,
					query
				};
			}
		}
		redirect(href) {
			dispatch$2(this, "history/redirect", { href });
		}
	};
	//#endregion
	Object.defineProperty(exports, "BrowserHistory", {
		enumerable: true,
		get: function() {
			return history_exports;
		}
	});
	Object.defineProperty(exports, "Switch", {
		enumerable: true,
		get: function() {
			return switch_exports;
		}
	});
	exports.fromHistory = fromHistory;
});
