// @unbndl/view/dist/view.umd.cjs
(function(global, factory) {
	typeof exports === "object" && typeof module !== "undefined" ? factory(exports) : typeof define === "function" && define.amd ? define(["exports"], factory) : (global = typeof globalThis !== "undefined" ? globalThis : global || self, factory(global.view = {}));
})(this, function(exports) {
	Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
	//#region src/source.ts
	var MappedSource = class {
		constructor(source, mapping) {
			this.initialized = false;
			this.origin = source;
			const entries = mapEntries(mapping).map(([t, s]) => [s, t]);
			this.inverse = Object.fromEntries(entries);
		}
		mapObservation(source) {
			const entries = Object.entries(source).map(([s, v]) => {
				const t = this.inverse[s];
				return t ? [t, v] : [];
			});
			return Object.fromEntries(entries.filter((pair) => pair.length > 0));
		}
		start(fn) {
			const sfn = (s, value) => {
				const t = this.inverse[s];
				fn(t, value);
				this.initialized = true;
			};
			return this.origin.start(sfn).then((obs) => this.initialized ? {} : this.mapObservation(obs));
		}
	};
	function mapEntries(mapping) {
		return Object.entries(mapping).map(([k, v]) => [k, v]);
	}
	//#endregion
	//#region ../html/dist/html.js
	var SignalEvent = class extends CustomEvent {
		constructor(eventType, signal) {
			super(eventType, {
				bubbles: true,
				cancelable: true,
				composed: true,
				detail: signal
			});
		}
	};
	function createEffect(fn, ...scope) {
		const effect = { execute() {
			console.log("▶️ Executing created effect:", scope, fn);
			fn(...scope.map((cx) => cx.open(effect)));
			scope.forEach((cx) => cx.close());
		} };
		effect.execute();
	}
	var Scheduler = class Scheduler {
		constructor() {
			this.signals = /* @__PURE__ */ new WeakMap();
			this.scheduled = /* @__PURE__ */ new WeakSet();
		}
		static {
			this.scheduler = new Scheduler();
		}
		subscribe(scope, key, effect) {
			let signals = this.signals.get(scope);
			if (!signals) {
				signals = /* @__PURE__ */ new Map();
				this.signals.set(scope, signals);
			}
			let signal = signals.get(key);
			if (!signal) {
				signal = /* @__PURE__ */ new Set();
				signals.set(key, signal);
			}
			signal.add(effect);
		}
		scheduleEffects(scope, key) {
			const signals = this.signals.get(scope);
			if (!signals) return;
			const signal = signals.get(key);
			if (signal) {
				for (const effect of signal) if (!this.scheduled.has(effect)) {
					this.scheduled.add(effect);
					setTimeout(() => {
						this.scheduled.delete(effect);
						effect.execute();
					});
				}
			}
		}
	};
	var EffectsManager = class {
		constructor() {
			this.running = [];
			this.eventType = "un-effect:change";
		}
		isRunning() {
			return this.running.length > 0;
		}
		push(effect) {
			this.running.push(effect);
		}
		pop() {
			this.running.pop();
		}
		current() {
			const len = this.running.length;
			return len ? this.running[len - 1] : void 0;
		}
		subscribe(key, scope) {
			const current = this.current();
			if (current) Scheduler.scheduler.subscribe(scope, key, current);
		}
		runEffects(key, scope) {
			console.log("⚙️ Running effects for signal", key, scope);
			Scheduler.scheduler.scheduleEffects(scope, key);
			if (this.host) {
				const evt = new SignalEvent(this.eventType, {
					property: key,
					value: scope[key]
				});
				this.host.dispatchEvent(evt);
			}
		}
		setHost(host, eventType) {
			this.host = host;
			if (eventType) this.eventType = eventType;
		}
	};
	var Context = class {
		static {
			this.CHANGE_EVENT_TYPE = "un-context:change";
		}
		constructor(init, adoptedContext) {
			if (adoptedContext) {
				this.manager = adoptedContext.manager;
				this.object = adoptedContext.object;
				this.proxy = adoptedContext.proxy;
				this.update(init);
			} else {
				this.manager = new EffectsManager();
				this.object = init;
				this.proxy = createContext(this.object, this.manager);
			}
		}
		get(prop) {
			return this.proxy[prop];
		}
		set(prop, value) {
			this.proxy[prop] = value;
		}
		toObject() {
			return this.proxy;
		}
		update(next) {
			Object.assign(this.proxy, next);
		}
		apply(mapFn) {
			this.update(mapFn(this.proxy));
		}
		createEffect(fn) {
			createEffect(fn, this);
		}
		setHost(host, eventType) {
			this.manager.setHost(host, eventType);
		}
		open(effect) {
			this.manager.push(effect);
			return this.proxy;
		}
		close() {
			this.manager.pop();
		}
	};
	function createContext(root, manager) {
		return new Proxy(root, {
			get: (subject, prop, receiver) => {
				const value = Reflect.get(subject, prop, receiver);
				if (manager.isRunning() && isObservable(value)) manager.subscribe(prop, subject);
				return value;
			},
			set: (subject, prop, newValue, receiver) => {
				const didSet = Reflect.set(subject, prop, newValue, receiver);
				if (didSet && isObservable(newValue)) manager.runEffects(prop, subject);
				return didSet;
			}
		});
	}
	function isObservable(value) {
		switch (typeof value) {
			case "object":
			case "number":
			case "string":
			case "symbol":
			case "boolean":
			case "undefined": return true;
			default: return false;
		}
	}
	function createScope(tuple) {
		return tuple.map((a) => typeof a === "undefined" ? null : new Context(a));
	}
	function createTemplate(fragment, render) {
		Object.assign(fragment, { render });
		return fragment;
	}
	function renderForEffects(original, effectors, ...scope) {
		const fragment = original.cloneNode(true);
		Array.from(effectors.entries()).forEach(([label, mutations]) => {
			const site = fragment.querySelector(`[data-${label}]`);
			if (site) mutations.forEach((fn) => fn(site, fragment, ...scope));
		});
		return fragment;
	}
	var TemplateParser = class TemplateParser {
		static {
			this.parser = new DOMParser();
		}
		constructor(docType) {
			this.docType = "text/html";
			this.plugins = [];
			if (docType) this.docType = docType;
		}
		use(plugin) {
			this.plugins = this.plugins.concat(plugin);
		}
		parse(template, params) {
			const postProcess = {};
			const stringToParse = template.map((s, i) => {
				if (i >= params.length) return [s];
				const param = params[i];
				const place = this.classifyPlace(i, template);
				const mutation = this.tryReplacements(place, param);
				if (mutation) {
					const post = postProcess[place.nodeLabel];
					if (post) post.push(mutation);
					else postProcess[place.nodeLabel] = [mutation];
					switch (place.kind) {
						case "attr value": return [s, `"" data-${place.nodeLabel}`];
						case "tag content": return [s, `data-${place.nodeLabel}`];
						case "element content": return [s, `<ins data-${place.nodeLabel}></ins>`];
					}
				} else {
					console.error("No match for template parameter", place, param);
					throw `Failed to render template parameter ${i} around ${s}`;
				}
				return [s];
			}).flat().join("");
			const doc = TemplateParser.parser.parseFromString(stringToParse, this.docType);
			const collection = doc.head.childElementCount ? doc.head.children : doc.body.children;
			const fragment = new DocumentFragment();
			fragment.replaceChildren(...collection);
			const effectors = /* @__PURE__ */ new Map();
			for (const label in postProcess) {
				const site = fragment.querySelector(`[data-${label}]`);
				if (site) postProcess[label].forEach((m) => {
					const effector = m.apply(site, fragment);
					if (effector) {
						let list = effectors.get(label);
						if (list) list.push(effector);
						else effectors.set(label, [effector]);
					}
				});
			}
			return createTemplate(fragment, (...scope) => renderForEffects(fragment, effectors, ...scope));
		}
		static {
			this.OPEN_RE = /<([a-zA-z][$a-zA-Z0-9.-]*)\s+[^>]*$/;
		}
		static {
			this.IN_TAG_RE = /^(\s+|[^<>]*|"[^"]*")*$/;
		}
		static {
			this.ATTR_RE = /([$.]?[a-zA-Z][$a-zA-Z0-9.-]*)=\s*$/;
		}
		static {
			this.CLOSE_RE = /[/]?>[^<]*$/;
		}
		classifyPlace(i, template) {
			let tagOpen = null;
			for (let prev = i; prev >= 0; prev--) {
				if (template[prev].match(TemplateParser.CLOSE_RE)) break;
				tagOpen = template[prev].match(TemplateParser.OPEN_RE);
				if (tagOpen) break;
				if (!template[prev].match(TemplateParser.IN_TAG_RE)) break;
			}
			if (tagOpen) {
				const tagAttr = template[i].match(TemplateParser.ATTR_RE);
				if (tagAttr) return {
					kind: "attr value",
					nodeLabel: `node${i}`,
					tagName: tagOpen[1],
					attrName: tagAttr[1]
				};
				return {
					kind: "tag content",
					nodeLabel: `node${i}`,
					tagName: tagOpen[1]
				};
			}
			return {
				kind: "element content",
				nodeLabel: `node${i}`
			};
		}
		tryReplacements(place, param) {
			const replacements = this.plugins;
			for (let i = 0; i < replacements.length; i++) {
				const sub = replacements[i];
				if (place.kind === sub.place && checkType(param, sub)) return sub.mutator(place, param);
			}
		}
	};
	function checkType(param, sub) {
		if (typeof sub.types === "function") return sub.types(param, sub);
		return sub.types.includes(typeof param);
	}
	var Mutation = class {
		constructor(place) {
			this.place = place;
		}
		apply(_site, _fragment) {
			throw "abstract method 'apply' called";
		}
	};
	var ElementContentMutation = class extends Mutation {
		constructor(place, content) {
			super(place);
			this.content = content;
		}
		apply(site, fragment) {
			(site.parentNode || fragment).replaceChild(this.content, site);
		}
	};
	var AttributeValueMutation = class extends Mutation {
		constructor(place, text) {
			super(place);
			this.text = text;
			this.name = place.attrName;
		}
		apply(site) {
			site.setAttribute(this.name, this.text);
		}
	};
	var ElementContentEffect = class extends Mutation {
		constructor(place, fn) {
			super(place);
			this.fn = fn;
		}
		apply(_site, _fragment) {
			const key = this.place.nodeLabel;
			return (site, fragment, ...scope) => {
				const start = new Comment(` <<< ${key} `);
				const end = new Comment(` >>> ${key} `);
				const placeholder = new DocumentFragment();
				placeholder.replaceChildren(start, end);
				(site.parentNode || fragment).replaceChild(placeholder, site);
				createEffect((...args) => {
					replaceElementContent(this.fn(...args), start, end);
				}, ...scope);
			};
		}
	};
	function replaceElementContent(value, start, end) {
		const parent = start.parentNode;
		if (!parent) throw new Error("No parent for placeholder");
		const valueToNode = (v) => {
			if (Array.isArray(v)) {
				const frag = new DocumentFragment();
				const nodes = v.map(valueToNode);
				frag.replaceChildren(...nodes);
				return frag;
			} else if (v instanceof Node) return v;
			else return new Text(v?.toString() || "");
		};
		const node = valueToNode(value);
		let p = start.nextSibling;
		while (p && p !== end) {
			const old = p;
			p = p.nextSibling;
			parent.removeChild(old);
		}
		if (node) parent.insertBefore(node, end);
	}
	var AttributeValueEffect = class extends Mutation {
		constructor(place, fn) {
			super(place);
			this.fn = fn;
			this.name = place.attrName;
		}
		apply(_site, _fragment) {
			return (site, _, ...scope) => createEffect((...args) => {
				replaceAttributeValue(this.fn(...args), site, this.name);
			}, ...scope);
		}
	};
	function replaceAttributeValue(value, site, attrName) {
		const special = attrName.match(/^([.$])(.+)$/);
		if (special) {
			const [_, pre, name] = special;
			switch (pre) {
				case ".":
					site[name] = value;
					break;
				case "$":
					if ("viewModel" in site && site.viewModel instanceof Context) site.viewModel.set(name, value);
					break;
			}
		} else switch (typeof value) {
			case "string":
				site.setAttribute(attrName, value);
				break;
			case "undefined":
			case "object":
			case "boolean":
				if (value) site.setAttribute(attrName, attrName);
				else site.removeAttribute(attrName);
				break;
			default: site.setAttribute(attrName, value?.toString());
		}
	}
	var TagReferenceEffect = class extends Mutation {
		constructor(place, fn) {
			super(place);
			this.fn = fn;
		}
		apply(_site, _fragment) {
			return (site, _, ...scope) => createEffect((...args) => {
				const value = this.fn(...args);
				if (typeof value === "function") value(site);
			}, ...scope);
		}
	};
	new TemplateParser().use([
		{
			place: "element content",
			types: [
				"string",
				"number",
				"bigint",
				"symbol",
				"boolean"
			],
			mutator: (place, value) => new ElementContentMutation(place, new Text(value?.toString() || ""))
		},
		{
			place: "attr value",
			types: [
				"string",
				"number",
				"bigint",
				"symbol"
			],
			mutator: (place, value) => new AttributeValueMutation(place, value?.toString() || "")
		},
		{
			place: "element content",
			types: (param) => param instanceof Node,
			mutator: (place, value) => new ElementContentMutation(place, value)
		},
		{
			place: "element content",
			types: (param) => Array.isArray(param),
			mutator: (place, value) => {
				const fragment = new DocumentFragment();
				const nodes = value.map((v) => v instanceof Node ? v : new Text(v?.toString() || ""));
				fragment.append(...nodes);
				return new ElementContentMutation(place, fragment);
			}
		},
		{
			place: "element content",
			types: ["function"],
			mutator: (place, param) => new ElementContentEffect(place, param)
		},
		{
			place: "attr value",
			types: ["function"],
			mutator: (place, param) => new AttributeValueEffect(place, param)
		},
		{
			place: "tag content",
			types: ["function"],
			mutator: (place, param) => new TagReferenceEffect(place, param)
		}
	]);
	//#endregion
	//#region src/view.ts
	function createView(html) {
		return html;
	}
	function createView2(html) {
		return html;
	}
	function createViewN(html) {
		return html;
	}
	function zipArgs(...lists) {
		const maxLength = lists.map((arr) => arr.length).reduce((a, b) => Math.max(a, b), 0);
		return Array.from(Array(maxLength).keys()).map((i) => lists.map((arr) => arr[i]));
	}
	function map(view, list) {
		return mapN(view, list);
	}
	function map2(view, ulist, vlist) {
		return mapN(view, ulist, vlist);
	}
	function mapN(view, ...lists) {
		if (!lists.length) return "";
		return zipArgs(...lists).map((tuple) => applyN(view, ...tuple)).flat();
	}
	function apply(view, t) {
		return typeof t === "undefined" ? "" : applyN(view, t);
	}
	function apply2(view, u, v) {
		return typeof u === "undefined" || typeof v === "undefined" ? "" : applyN(view, u, v);
	}
	function applyN(view, ...tuple) {
		if (!tuple) return "";
		const scope = createScope(tuple);
		return view.render(...scope);
	}
	var View = {
		apply,
		apply2,
		applyN,
		map,
		map2,
		mapN
	};
	//#endregion
	//#region src/viewModel.ts
	var ViewModel = class extends Context {
		constructor(init, adoptedContext) {
			super(init, adoptedContext);
		}
		get $() {
			return this.toObject();
		}
		with(source, ...keys) {
			const mapping = Object.fromEntries(keys.map((k) => [k, k]));
			return this.merge(new MappedSource(source, mapping));
		}
		withCalculated(source, mapping) {
			return this.merge(new MappedSource(source, mapping));
		}
		withRenamed(source, renaming) {
			return this.merge(new MappedSource(source, renaming));
		}
		merge(source) {
			if (source) {
				const initialized = /* @__PURE__ */ new Map();
				source.start((name, value) => {
					this.set(name, value);
					initialized.set(name, true);
				}).then((firstObservation) => {
					Object.keys(firstObservation).filter((k) => !initialized.get(k)).forEach((k) => {
						this.set(k, firstObservation[k]);
					});
				});
			}
			return this;
		}
		render(template) {
			return template.render(this);
		}
	};
	function createViewModel(init) {
		if (init === void 0) return new ViewModel({});
		else return new ViewModel(init);
	}
	//#endregion
	//#region src/fromAttributes.ts
	function fromAttributes(subject) {
		return new FromAttributes(subject);
	}
	var FromAttributes = class {
		constructor(subject) {
			this.subject = subject;
		}
		start(fn) {
			const element = this.subject;
			return new Promise((resolve, _reject) => {
				const init = {};
				const attributes = element.attributes;
				for (const attr of attributes) init[attr.name] = attr.value;
				resolve(init);
				new MutationObserver(effectChanges).observe(element, { attributes: true });
			});
			function effectChanges(mutations) {
				mutations.forEach((mut) => {
					const name = mut.attributeName;
					fn(name, element.getAttribute(name));
				});
			}
		}
	};
	//#endregion
	//#region src/fromInputs.ts
	function fromInputs(subject) {
		return new FromInputs(subject);
	}
	var FromInputs = class {
		constructor(subject) {
			this.subject = subject;
		}
		start(fn) {
			this.subject.addEventListener("change", (event) => {
				const input = event.target;
				if (input) {
					const name = input.name;
					const value = input.value;
					fn(name, value);
				}
			});
			return new Promise((_resolve, _reject) => {});
		}
	};
	//#endregion
	exports.MappedSource = MappedSource;
	exports.View = View;
	exports.ViewModel = ViewModel;
	exports.createView = createView;
	exports.createView2 = createView2;
	exports.createViewModel = createViewModel;
	exports.createViewN = createViewN;
	exports.fromAttributes = fromAttributes;
	exports.fromInputs = fromInputs;
});
