// @unbndl/service/src/service.ts
import { Context } from "@unbndl/html";
import { Provider } from "./provider.ts";
import * as Message from "./message.ts";
import { Update } from "./update.ts";

type None = Message.None;

export class Service<
  Msg extends Message.Base,
  T extends object
> {
  _context: Context<T>;
  _update: Update<Msg, T>;
  _eventType: string;
  _running: boolean;
  _pending: Array<Msg> = [];

  constructor(
    update: Update<Msg, T>,
    context: Context<T>,
    eventType = "service:message",
    autostart = true
  ) {
    this._context = context;
    this._update = update;
    this._eventType = eventType;
    this._running = autostart;
  }

  attach(host: Provider<T>) {
    host.addEventListener(this._eventType, (ev: Event) => {
      ev.stopPropagation();
      const message = (ev as Message.Dispatch<Msg>).detail;
      this.consume(message);
    });
  }

  start() {
    if (!this._running) {
      console.log(`Starting ${this._eventType} service`);
      this._running = true;
      this._pending.forEach((msg) => this.process(msg));
    }
  }

  consume(message: Msg | None) {
    // console.log("Consume", message, this._running);
    if (message.length !== 0 ) {
      const trueMsg = message as Msg;
      if (this._running) {
        this.process(trueMsg);
      } else {
        console.log(
           `📥 Queueing ${this._eventType} message`,
           message
         );
        this._pending.push(trueMsg);
      }
    }
  }

  process(message: Msg) {
    const reactions = message[2];
    let next: T | Message.Async<T, Msg> | undefined;

    try {
      console.log(
        `📤 Processing ${this._eventType} message`,
        message
      );
      next = this._update(
        message,
        this._context.toObject()
      );
      console.log("⏭️ next state: ", next);
    } catch (err: unknown) {
      console.log(
        `‼️ Error while processing ${message[0]} message:`, err
      );
      if (reactions && reactions.onFailure) {
        const error = err instanceof Error ? err
          : new Error(err as string);
        reactions.onFailure(error);
      }
    }

    if (!next) return;

    if (!Array.isArray(next)) {
      this._context.update(next);
    } else {
      const [now, ...later] = next;
      later.forEach((promise) =>
        promise.then((command: Msg | None) =>
          this.consume(command)
        ).catch((err: Error) => {
          console.log(
            `‼️ Error while processing later ${this._eventType} command:`, err.message
          );
          if (reactions && reactions.onFailure) {
            reactions.onFailure(err);
          }
        })
      );
      this._context.update(now);
    }

    if (reactions && reactions.onSuccess) reactions.onSuccess();
  }
}
