// @unbndl/switch/src/history.ts
import { Context } from "@unbndl/html";
import { Message, Provider, Service } from "@unbndl/service";

const HISTORY_CONTEXT_DEFAULT = "context:history";

interface HistoryModel {
  location: Location;
  state: object | undefined;
}

type HistoryMsg =
  | ["history/navigate", {
  href: string;
  state?: object;
}]
  | ["history/redirect", {
  href: string;
  state?: object;
}];

const cloneCurrentLocation = () => Object.assign({}, document.location);

class HistoryService extends Service<HistoryMsg, HistoryModel> {
  static EVENT_TYPE = "history:message";

  constructor(context: Context<HistoryModel>) {
    super(
      (msg, model) => this.update(msg, model),
      context,
      HistoryService.EVENT_TYPE
    );
  }

  update(
    message: HistoryMsg,
    model: HistoryModel
  ):
    HistoryModel {
    switch (message[0]) {
      case "history/navigate": {
        const { href, state } = message[1];
        history.pushState(state, "", href);
        return {
          ...model,
          location: cloneCurrentLocation(),
          state: history.state
        };
      }
      case "history/redirect": {
        const { href, state } = message[1];
        history.replaceState(state, "", href);
        return {
          ...model,
          location: cloneCurrentLocation(),
          state: history.state
        };
      }
    }
    return model;
  }
}

class HistoryProvider extends Provider<HistoryModel> {
  constructor() {
    super(
      {
        location: cloneCurrentLocation(),
        state: {}
      },
      HISTORY_CONTEXT_DEFAULT
    );

    // console.log(
    //   "📍 History location:",
    //   this.context.get("location")
    // );

    this.addEventListener("click", (event: MouseEvent) => {
      const linkTarget = originalLinkTarget(event);
      if (linkTarget) {
        // It's a left click on an <a href=...>.
        const url = new URL(linkTarget.href);

        const location = this.context.get(
          "location"
        ) as Location;
        if (
          location &&
          url.origin === location.origin &&
          url.pathname.startsWith(this.base || "/")
        ) {
          // console.log("Preventing Click Event on <A>", event);
          event.preventDefault();
          dispatch(linkTarget, "history/navigate", {
            href: url.pathname + url.search
          });
        }
      }
    });

    window.addEventListener("popstate", (event) => {
      // console.log("Popstate", event.state);
      this.context.update({
        location: cloneCurrentLocation(),
        state: event.state
      });
    });
  }

  get base() {
    return this.getAttribute("base") || undefined;
  }

  connectedCallback() {
    const service = new HistoryService(this.context);
    service.attach(this);
  }

  attributeChangedCallback() {
  }
}

function originalLinkTarget(
  event: MouseEvent
): HTMLAnchorElement | undefined {
  const current = event.currentTarget;
  const isLink = (el: EventTarget) =>
    (el as HTMLElement).tagName == "A" &&
    (el as HTMLAnchorElement).href;

  if (event.button !== 0) return undefined;

  if (event.composed) {
    const path = event.composedPath();
    const target = path.find(isLink);
    return target ? (target as HTMLAnchorElement) : undefined;
  } else {
    for (
      let target: HTMLElement | null =
        event.target as HTMLElement;
      target;
      target === current
        ? null
        : (target as HTMLElement).parentElement
    ) {
      if (isLink(target)) return target as HTMLAnchorElement;
    }
    return undefined;
  }
}


const dispatch = Message.dispatcher<HistoryMsg>(
  HistoryService.EVENT_TYPE
);

export {
  HISTORY_CONTEXT_DEFAULT as CONTEXT_DEFAULT,
  HistoryProvider as Provider,
  HistoryService as Service,
  type HistoryModel as Model,
  type HistoryMsg as Message,
  dispatch
};
