// @unbndl/store/dist/store.js
//#region \0rolldown/runtime.js
var __defProp$2 = Object.defineProperty;
var __exportAll$2 = (all, no_symbols) => {
	let target = {};
	for (var name in all) __defProp$2(target, name, {
		get: all[name],
		enumerable: true
	});
	if (!no_symbols) __defProp$2(target, Symbol.toStringTag, { value: "Module" });
	return target;
};
//#endregion
//#region ../auth/dist/auth.js
var InvalidTokenError = class extends Error {};
InvalidTokenError.prototype.name = "InvalidTokenError";
function b64DecodeUnicode(str) {
	return decodeURIComponent(atob(str).replace(/(.)/g, (m, p) => {
		let code = p.charCodeAt(0).toString(16).toUpperCase();
		if (code.length < 2) code = "0" + code;
		return "%" + code;
	}));
}
function base64UrlDecode(str) {
	let output = str.replace(/-/g, "+").replace(/_/g, "/");
	switch (output.length % 4) {
		case 0: break;
		case 2:
			output += "==";
			break;
		case 3:
			output += "=";
			break;
		default: throw new Error("base64 string is not of the correct length");
	}
	try {
		return b64DecodeUnicode(output);
	} catch (err) {
		return atob(output);
	}
}
function jwtDecode(token, options) {
	if (typeof token !== "string") throw new InvalidTokenError("Invalid token specified: must be a string");
	options || (options = {});
	const pos = options.header === true ? 0 : 1;
	const part = token.split(".")[pos];
	if (typeof part !== "string") throw new InvalidTokenError(`Invalid token specified: missing part #${pos + 1}`);
	let decoded;
	try {
		decoded = base64UrlDecode(part);
	} catch (e) {
		throw new InvalidTokenError(`Invalid token specified: invalid base64 for part #${pos + 1} (${e.message})`);
	}
	try {
		return JSON.parse(decoded);
	} catch (e) {
		throw new InvalidTokenError(`Invalid token specified: invalid json for part #${pos + 1} (${e.message})`);
	}
}
var __defProp$1 = Object.defineProperty;
var __exportAll$1 = (all, no_symbols) => {
	let target = {};
	for (var name in all) __defProp$1(target, name, {
		get: all[name],
		enumerable: true
	});
	if (!no_symbols) __defProp$1(target, Symbol.toStringTag, { value: "Module" });
	return target;
};
var message_exports$1 = /* @__PURE__ */ __exportAll$1({
	Dispatch: () => Dispatch$1,
	None: () => None$1,
	dispatch: () => dispatch$1$1,
	dispatcher: () => dispatcher$1
});
var None$1 = [];
var Dispatch$1 = class extends CustomEvent {
	constructor(msg, eventType = "un:message") {
		super(eventType, {
			bubbles: true,
			composed: true,
			detail: msg
		});
	}
};
function dispatcher$1(eventType = "un:message") {
	return (target, ...msg) => target.dispatchEvent(new Dispatch$1(msg, eventType));
}
var dispatch$1$1 = dispatcher$1();
var SignalEvent$2 = class extends CustomEvent {
	constructor(eventType, signal) {
		super(eventType, {
			bubbles: true,
			cancelable: true,
			composed: true,
			detail: signal
		});
	}
};
function createEffect$2(fn, ...scope) {
	const effect = { execute() {
		console.log("▶️ Executing created effect:", scope, fn);
		fn(...scope.map((cx) => cx.open(effect)));
		scope.forEach((cx) => cx.close());
	} };
	effect.execute();
}
var Scheduler$2 = class Scheduler {
	constructor() {
		this.signals = /* @__PURE__ */ new WeakMap();
		this.scheduled = /* @__PURE__ */ new WeakSet();
	}
	static {
		this.scheduler = new Scheduler();
	}
	subscribe(scope, key, effect) {
		let signals = this.signals.get(scope);
		if (!signals) {
			signals = /* @__PURE__ */ new Map();
			this.signals.set(scope, signals);
		}
		let signal = signals.get(key);
		if (!signal) {
			signal = /* @__PURE__ */ new Set();
			signals.set(key, signal);
		}
		signal.add(effect);
	}
	scheduleEffects(scope, key) {
		const signals = this.signals.get(scope);
		if (!signals) return;
		const signal = signals.get(key);
		if (signal) {
			for (const effect of signal) if (!this.scheduled.has(effect)) {
				this.scheduled.add(effect);
				setTimeout(() => {
					this.scheduled.delete(effect);
					effect.execute();
				});
			}
		}
	}
};
var EffectsManager$2 = class {
	constructor() {
		this.running = [];
		this.eventType = "un-effect:change";
	}
	isRunning() {
		return this.running.length > 0;
	}
	push(effect) {
		this.running.push(effect);
	}
	pop() {
		this.running.pop();
	}
	current() {
		const len = this.running.length;
		return len ? this.running[len - 1] : void 0;
	}
	subscribe(key, scope) {
		const current = this.current();
		if (current) Scheduler$2.scheduler.subscribe(scope, key, current);
	}
	runEffects(key, scope) {
		console.log("⚙️ Running effects for signal", key, scope);
		Scheduler$2.scheduler.scheduleEffects(scope, key);
		if (this.host) {
			const evt = new SignalEvent$2(this.eventType, {
				property: key,
				value: scope[key]
			});
			this.host.dispatchEvent(evt);
		}
	}
	setHost(host, eventType) {
		this.host = host;
		if (eventType) this.eventType = eventType;
	}
};
var Context$2 = class {
	static {
		this.CHANGE_EVENT_TYPE = "un-context:change";
	}
	constructor(init, adoptedContext) {
		if (adoptedContext) {
			this.manager = adoptedContext.manager;
			this.object = adoptedContext.object;
			this.proxy = adoptedContext.proxy;
			this.update(init);
		} else {
			this.manager = new EffectsManager$2();
			this.object = init;
			this.proxy = createContext$2(this.object, this.manager);
		}
	}
	get(prop) {
		return this.proxy[prop];
	}
	set(prop, value) {
		this.proxy[prop] = value;
	}
	toObject() {
		return this.proxy;
	}
	update(next) {
		Object.assign(this.proxy, next);
	}
	apply(mapFn) {
		this.update(mapFn(this.proxy));
	}
	createEffect(fn) {
		createEffect$2(fn, this);
	}
	setHost(host, eventType) {
		this.manager.setHost(host, eventType);
	}
	open(effect) {
		this.manager.push(effect);
		return this.proxy;
	}
	close() {
		this.manager.pop();
	}
};
function createContext$2(root, manager) {
	return new Proxy(root, {
		get: (subject, prop, receiver) => {
			const value = Reflect.get(subject, prop, receiver);
			if (manager.isRunning() && isObservable$2(value)) manager.subscribe(prop, subject);
			return value;
		},
		set: (subject, prop, newValue, receiver) => {
			const didSet = Reflect.set(subject, prop, newValue, receiver);
			if (didSet && isObservable$2(newValue)) manager.runEffects(prop, subject);
			return didSet;
		}
	});
}
function isObservable$2(value) {
	switch (typeof value) {
		case "object":
		case "number":
		case "string":
		case "symbol":
		case "boolean":
		case "undefined": return true;
		default: return false;
	}
}
var DirectEffect$1 = class {
	constructor(fn, ...scope) {
		this.effectFn = () => fn(...scope);
	}
	execute() {
		this.effectFn();
	}
};
function createTemplate$2(fragment, render) {
	Object.assign(fragment, { render });
	return fragment;
}
function renderForEffects$2(original, effectors, ...scope) {
	const fragment = original.cloneNode(true);
	Array.from(effectors.entries()).forEach(([label, mutations]) => {
		const site = fragment.querySelector(`[data-${label}]`);
		if (site) mutations.forEach((fn) => fn(site, fragment, ...scope));
	});
	return fragment;
}
var TemplateParser$2 = class TemplateParser {
	static {
		this.parser = new DOMParser();
	}
	constructor(docType) {
		this.docType = "text/html";
		this.plugins = [];
		if (docType) this.docType = docType;
	}
	use(plugin) {
		this.plugins = this.plugins.concat(plugin);
	}
	parse(template, params) {
		const postProcess = {};
		const stringToParse = template.map((s, i) => {
			if (i >= params.length) return [s];
			const param = params[i];
			const place = this.classifyPlace(i, template);
			const mutation = this.tryReplacements(place, param);
			if (mutation) {
				const post = postProcess[place.nodeLabel];
				if (post) post.push(mutation);
				else postProcess[place.nodeLabel] = [mutation];
				switch (place.kind) {
					case "attr value": return [s, `"" data-${place.nodeLabel}`];
					case "tag content": return [s, `data-${place.nodeLabel}`];
					case "element content": return [s, `<ins data-${place.nodeLabel}></ins>`];
				}
			} else {
				console.error("No match for template parameter", place, param);
				throw `Failed to render template parameter ${i} around ${s}`;
			}
			return [s];
		}).flat().join("");
		const doc = TemplateParser.parser.parseFromString(stringToParse, this.docType);
		const collection = doc.head.childElementCount ? doc.head.children : doc.body.children;
		const fragment = new DocumentFragment();
		fragment.replaceChildren(...collection);
		const effectors = /* @__PURE__ */ new Map();
		for (const label in postProcess) {
			const site = fragment.querySelector(`[data-${label}]`);
			if (site) postProcess[label].forEach((m) => {
				const effector = m.apply(site, fragment);
				if (effector) {
					let list = effectors.get(label);
					if (list) list.push(effector);
					else effectors.set(label, [effector]);
				}
			});
		}
		return createTemplate$2(fragment, (...scope) => renderForEffects$2(fragment, effectors, ...scope));
	}
	static {
		this.OPEN_RE = /<([a-zA-z][$a-zA-Z0-9.-]*)\s+[^>]*$/;
	}
	static {
		this.IN_TAG_RE = /^(\s+|[^<>]*|"[^"]*")*$/;
	}
	static {
		this.ATTR_RE = /([$.]?[a-zA-Z][$a-zA-Z0-9.-]*)=\s*$/;
	}
	static {
		this.CLOSE_RE = /[/]?>[^<]*$/;
	}
	classifyPlace(i, template) {
		let tagOpen = null;
		for (let prev = i; prev >= 0; prev--) {
			if (template[prev].match(TemplateParser.CLOSE_RE)) break;
			tagOpen = template[prev].match(TemplateParser.OPEN_RE);
			if (tagOpen) break;
			if (!template[prev].match(TemplateParser.IN_TAG_RE)) break;
		}
		if (tagOpen) {
			const tagAttr = template[i].match(TemplateParser.ATTR_RE);
			if (tagAttr) return {
				kind: "attr value",
				nodeLabel: `node${i}`,
				tagName: tagOpen[1],
				attrName: tagAttr[1]
			};
			return {
				kind: "tag content",
				nodeLabel: `node${i}`,
				tagName: tagOpen[1]
			};
		}
		return {
			kind: "element content",
			nodeLabel: `node${i}`
		};
	}
	tryReplacements(place, param) {
		const replacements = this.plugins;
		for (let i = 0; i < replacements.length; i++) {
			const sub = replacements[i];
			if (place.kind === sub.place && checkType$2(param, sub)) return sub.mutator(place, param);
		}
	}
};
function checkType$2(param, sub) {
	if (typeof sub.types === "function") return sub.types(param, sub);
	return sub.types.includes(typeof param);
}
var Mutation$2 = class {
	constructor(place) {
		this.place = place;
	}
	apply(_site, _fragment) {
		throw "abstract method 'apply' called";
	}
};
var ElementContentMutation$2 = class extends Mutation$2 {
	constructor(place, content) {
		super(place);
		this.content = content;
	}
	apply(site, fragment) {
		(site.parentNode || fragment).replaceChild(this.content, site);
	}
};
var AttributeValueMutation$2 = class extends Mutation$2 {
	constructor(place, text) {
		super(place);
		this.text = text;
		this.name = place.attrName;
	}
	apply(site) {
		site.setAttribute(this.name, this.text);
	}
};
var ElementContentEffect$2 = class extends Mutation$2 {
	constructor(place, fn) {
		super(place);
		this.fn = fn;
	}
	apply(_site, _fragment) {
		const key = this.place.nodeLabel;
		return (site, fragment, ...scope) => {
			const start = new Comment(` <<< ${key} `);
			const end = new Comment(` >>> ${key} `);
			const placeholder = new DocumentFragment();
			placeholder.replaceChildren(start, end);
			(site.parentNode || fragment).replaceChild(placeholder, site);
			createEffect$2((...args) => {
				replaceElementContent$2(this.fn(...args), start, end);
			}, ...scope);
		};
	}
};
function replaceElementContent$2(value, start, end) {
	const parent = start.parentNode;
	if (!parent) throw new Error("No parent for placeholder");
	const valueToNode = (v) => {
		if (Array.isArray(v)) {
			const frag = new DocumentFragment();
			const nodes = v.map(valueToNode);
			frag.replaceChildren(...nodes);
			return frag;
		} else if (v instanceof Node) return v;
		else return new Text(v?.toString() || "");
	};
	const node = valueToNode(value);
	let p = start.nextSibling;
	while (p && p !== end) {
		const old = p;
		p = p.nextSibling;
		parent.removeChild(old);
	}
	if (node) parent.insertBefore(node, end);
}
var AttributeValueEffect$2 = class extends Mutation$2 {
	constructor(place, fn) {
		super(place);
		this.fn = fn;
		this.name = place.attrName;
	}
	apply(_site, _fragment) {
		return (site, _, ...scope) => createEffect$2((...args) => {
			replaceAttributeValue$2(this.fn(...args), site, this.name);
		}, ...scope);
	}
};
function replaceAttributeValue$2(value, site, attrName) {
	const special = attrName.match(/^([.$])(.+)$/);
	if (special) {
		const [_, pre, name] = special;
		switch (pre) {
			case ".":
				site[name] = value;
				break;
			case "$":
				if ("viewModel" in site && site.viewModel instanceof Context$2) site.viewModel.set(name, value);
				break;
		}
	} else switch (typeof value) {
		case "string":
			site.setAttribute(attrName, value);
			break;
		case "undefined":
		case "object":
		case "boolean":
			if (value) site.setAttribute(attrName, attrName);
			else site.removeAttribute(attrName);
			break;
		default: site.setAttribute(attrName, value?.toString());
	}
}
var TagReferenceEffect$2 = class extends Mutation$2 {
	constructor(place, fn) {
		super(place);
		this.fn = fn;
	}
	apply(_site, _fragment) {
		return (site, _, ...scope) => createEffect$2((...args) => {
			const value = this.fn(...args);
			if (typeof value === "function") value(site);
		}, ...scope);
	}
};
new TemplateParser$2().use([
	{
		place: "element content",
		types: [
			"string",
			"number",
			"bigint",
			"symbol",
			"boolean"
		],
		mutator: (place, value) => new ElementContentMutation$2(place, new Text(value?.toString() || ""))
	},
	{
		place: "attr value",
		types: [
			"string",
			"number",
			"bigint",
			"symbol"
		],
		mutator: (place, value) => new AttributeValueMutation$2(place, value?.toString() || "")
	},
	{
		place: "element content",
		types: (param) => param instanceof Node,
		mutator: (place, value) => new ElementContentMutation$2(place, value)
	},
	{
		place: "element content",
		types: (param) => Array.isArray(param),
		mutator: (place, value) => {
			const fragment = new DocumentFragment();
			const nodes = value.map((v) => v instanceof Node ? v : new Text(v?.toString() || ""));
			fragment.append(...nodes);
			return new ElementContentMutation$2(place, fragment);
		}
	},
	{
		place: "element content",
		types: ["function"],
		mutator: (place, param) => new ElementContentEffect$2(place, param)
	},
	{
		place: "attr value",
		types: ["function"],
		mutator: (place, param) => new AttributeValueEffect$2(place, param)
	},
	{
		place: "tag content",
		types: ["function"],
		mutator: (place, param) => new TagReferenceEffect$2(place, param)
	}
]);
var Provider$1 = class Provider extends HTMLElement {
	static {
		this.DISCOVERY_EVENT = "un-provider:discover";
	}
	static {
		this.REGISTRY_EVENT = "un-provider:register";
	}
	static {
		this.CHANGE_EVENT = "un-provider:change";
	}
	static {
		document.addEventListener(Provider.DISCOVERY_EVENT, (event) => {
			const [contextLabel, respondFn] = event.detail;
			const provider = registeredProvider$1(contextLabel);
			if (provider) respondFn(provider);
		});
		document.addEventListener(Provider.REGISTRY_EVENT, (event) => {
			const [contextLabel, provider] = event.detail;
			registerProvider$1(contextLabel, provider);
		});
	}
	constructor(init, label) {
		super();
		this.contextLabel = label;
		this.context = new Context$2(init);
		this.context.setHost(this, Provider.CHANGE_EVENT);
		this.addEventListener(Provider.DISCOVERY_EVENT, (event) => {
			const [contextLabel, respondFn] = event.detail;
			if (contextLabel === this.contextLabel) {
				event.stopPropagation();
				respondFn(this);
			}
		});
		const registryEvent = new CustomEvent(Provider.REGISTRY_EVENT, {
			bubbles: true,
			composed: true,
			detail: [this.contextLabel, this]
		});
		this.dispatchEvent(registryEvent);
	}
	attach(observer) {
		this.addEventListener(Provider.CHANGE_EVENT, observer);
		return this.context.toObject();
	}
	detach(observer) {
		this.removeEventListener(Provider.CHANGE_EVENT, observer);
	}
};
function discover$1(observer, contextLabel) {
	return new Promise((resolve, reject) => {
		const discoveryEvent = new CustomEvent(Provider$1.DISCOVERY_EVENT, {
			bubbles: true,
			composed: true,
			detail: [contextLabel, (provider) => provider ? resolve(provider) : reject()]
		});
		if (observer.isConnected) observer.dispatchEvent(discoveryEvent);
		else document.dispatchEvent(discoveryEvent);
	});
}
var registry$1 = {};
function registerProvider$1(label, provider) {
	registry$1[label] = provider;
}
function registeredProvider$1(label) {
	return registry$1[label];
}
var Observer$1 = class {
	constructor(contextLabel) {
		this.contextLabel = contextLabel;
	}
	observe(from, fn) {
		return new Promise((resolve, reject) => {
			if (this.provider) resolve(this.attachObserver(fn));
			else discover$1(from, this.contextLabel).then((provider) => {
				this.provider = provider;
				resolve(this.attachObserver(fn));
			}).catch((err) => reject(err));
		});
	}
	attachObserver(fn) {
		const init = this.provider.attach((ev) => {
			const { property, value } = ev.detail;
			const effect = new DirectEffect$1(fn, {
				property,
				value
			});
			console.log("📡Signal received:", property, value);
			effect.execute();
		});
		console.log("👀 Initial observation:", init);
		return init;
	}
};
var Service$1 = class {
	constructor(update, context, eventType = "service:message", autostart = true) {
		this._pending = [];
		this._context = context;
		this._update = update;
		this._eventType = eventType;
		this._running = autostart;
	}
	attach(host) {
		host.addEventListener(this._eventType, (ev) => {
			ev.stopPropagation();
			const message = ev.detail;
			this.consume(message);
		});
	}
	start() {
		if (!this._running) {
			console.log(`Starting ${this._eventType} service`);
			this._running = true;
			this._pending.forEach((msg) => this.process(msg));
		}
	}
	consume(message) {
		if (message.length !== 0) {
			const trueMsg = message;
			if (this._running) this.process(trueMsg);
			else {
				console.log(`📥 Queueing ${this._eventType} message`, message);
				this._pending.push(trueMsg);
			}
		}
	}
	process(message) {
		console.log(`📤 Processing ${this._eventType} message`, message);
		const next = this._update(message, this._context.toObject());
		console.log("⏭️ next state: ", next);
		if (!Array.isArray(next)) this._context.update(next);
		else {
			const [now, ...later] = next;
			later.forEach((promise) => promise.then((command) => this.consume(command)));
			this._context.update(now);
		}
	}
};
function fromService$1(target, contextLabel) {
	return new FromService$1(target, contextLabel);
}
var FromService$1 = class {
	constructor(client, contextLabel) {
		this.client = client;
		this.observer = new Observer$1(contextLabel);
	}
	start(fn) {
		return this.observer.observe(this.client, (s) => {
			fn(s.property, s.value);
		});
	}
};
var AUTH_CONTEXT_DEFAULT = "context:auth";
var APIUser = class APIUser {
	static {
		this.TOKEN_KEY = "un-auth:token";
	}
	constructor(username) {
		this.authenticated = false;
		this.username = username || "anonymous";
	}
	static deauthenticate(user) {
		user.authenticated = false;
		user.username = "anonymous";
		localStorage.removeItem(APIUser.TOKEN_KEY);
		return user;
	}
};
var AuthenticatedUser = class AuthenticatedUser extends APIUser {
	constructor(token) {
		super();
		const jsonPayload = jwtDecode(token);
		console.log("Token payload", jsonPayload);
		this.token = token;
		this.authenticated = true;
		this.username = jsonPayload.username;
	}
	static authenticate(token) {
		const authenticatedUser = new AuthenticatedUser(token);
		localStorage.setItem(APIUser.TOKEN_KEY, token);
		return authenticatedUser;
	}
	static authenticateFromLocalStorage() {
		const priorToken = localStorage.getItem(APIUser.TOKEN_KEY);
		return priorToken ? AuthenticatedUser.authenticate(priorToken) : new APIUser();
	}
};
var AuthService = class AuthService extends Service$1 {
	static {
		this.EVENT_TYPE = "auth:message";
	}
	constructor(context, redirectForLogin) {
		super((msg, model) => this.update(msg, model), context, AuthService.EVENT_TYPE);
		this._redirectForLogin = redirectForLogin;
	}
	update(message, model) {
		console.log("Auth update", message);
		switch (message[0]) {
			case "auth/signin":
				const { token, redirect } = message[1];
				return [signIn(token), redirectLater(redirect)];
			case "auth/signout": return [signOut(model), redirectLater(this._redirectForLogin)];
			case "auth/redirect":
				redirectNow(this._redirectForLogin, { next: window.location.href });
				break;
			default:
				const unhandled = message[0];
				throw new Error(`Unhandled Auth message "${unhandled}"`);
		}
		return model;
	}
};
message_exports$1.dispatcher(AuthService.EVENT_TYPE);
function redirectLater(redirect, query = {}) {
	return new Promise((resolve) => {
		redirectNow(redirect, query);
		resolve(message_exports$1.None);
	});
}
function redirectNow(redirect, query = {}) {
	if (redirect) {
		const base = window.location.href;
		const target = new URL(redirect, base);
		Object.entries(query).forEach(([k, v]) => target.searchParams.set(k, v));
		console.log("Auth redirecting to ", redirect);
		window.history.pushState({}, "", redirect);
		window.history.go();
	}
}
function signIn(token) {
	const { authenticated, username } = AuthenticatedUser.authenticate(token);
	return {
		authenticated,
		username,
		token
	};
}
function signOut(model) {
	const { authenticated, username } = APIUser.deauthenticate(new APIUser(model.username));
	return {
		username,
		authenticated,
		token: void 0
	};
}
function fromAuth(target, contextLabel = AUTH_CONTEXT_DEFAULT) {
	return fromService$1(target, contextLabel);
}
//#endregion
//#region ../service/dist/service.js
var __defProp = Object.defineProperty;
var __exportAll = (all, no_symbols) => {
	let target = {};
	for (var name in all) __defProp(target, name, {
		get: all[name],
		enumerable: true
	});
	if (!no_symbols) __defProp(target, Symbol.toStringTag, { value: "Module" });
	return target;
};
var message_exports = /* @__PURE__ */ __exportAll({
	Dispatch: () => Dispatch,
	None: () => None,
	dispatch: () => dispatch$1,
	dispatcher: () => dispatcher
});
var None = [];
var Dispatch = class extends CustomEvent {
	constructor(msg, eventType = "un:message") {
		super(eventType, {
			bubbles: true,
			composed: true,
			detail: msg
		});
	}
};
function dispatcher(eventType = "un:message") {
	return (target, ...msg) => target.dispatchEvent(new Dispatch(msg, eventType));
}
var dispatch$1 = dispatcher();
var SignalEvent$1 = class extends CustomEvent {
	constructor(eventType, signal) {
		super(eventType, {
			bubbles: true,
			cancelable: true,
			composed: true,
			detail: signal
		});
	}
};
function createEffect$1(fn, ...scope) {
	const effect = { execute() {
		console.log("▶️ Executing created effect:", scope, fn);
		fn(...scope.map((cx) => cx.open(effect)));
		scope.forEach((cx) => cx.close());
	} };
	effect.execute();
}
var Scheduler$1 = class Scheduler {
	constructor() {
		this.signals = /* @__PURE__ */ new WeakMap();
		this.scheduled = /* @__PURE__ */ new WeakSet();
	}
	static {
		this.scheduler = new Scheduler();
	}
	subscribe(scope, key, effect) {
		let signals = this.signals.get(scope);
		if (!signals) {
			signals = /* @__PURE__ */ new Map();
			this.signals.set(scope, signals);
		}
		let signal = signals.get(key);
		if (!signal) {
			signal = /* @__PURE__ */ new Set();
			signals.set(key, signal);
		}
		signal.add(effect);
	}
	scheduleEffects(scope, key) {
		const signals = this.signals.get(scope);
		if (!signals) return;
		const signal = signals.get(key);
		if (signal) {
			for (const effect of signal) if (!this.scheduled.has(effect)) {
				this.scheduled.add(effect);
				setTimeout(() => {
					this.scheduled.delete(effect);
					effect.execute();
				});
			}
		}
	}
};
var EffectsManager$1 = class {
	constructor() {
		this.running = [];
		this.eventType = "un-effect:change";
	}
	isRunning() {
		return this.running.length > 0;
	}
	push(effect) {
		this.running.push(effect);
	}
	pop() {
		this.running.pop();
	}
	current() {
		const len = this.running.length;
		return len ? this.running[len - 1] : void 0;
	}
	subscribe(key, scope) {
		const current = this.current();
		if (current) Scheduler$1.scheduler.subscribe(scope, key, current);
	}
	runEffects(key, scope) {
		console.log("⚙️ Running effects for signal", key, scope);
		Scheduler$1.scheduler.scheduleEffects(scope, key);
		if (this.host) {
			const evt = new SignalEvent$1(this.eventType, {
				property: key,
				value: scope[key]
			});
			this.host.dispatchEvent(evt);
		}
	}
	setHost(host, eventType) {
		this.host = host;
		if (eventType) this.eventType = eventType;
	}
};
var Context$1 = class {
	static {
		this.CHANGE_EVENT_TYPE = "un-context:change";
	}
	constructor(init, adoptedContext) {
		if (adoptedContext) {
			this.manager = adoptedContext.manager;
			this.object = adoptedContext.object;
			this.proxy = adoptedContext.proxy;
			this.update(init);
		} else {
			this.manager = new EffectsManager$1();
			this.object = init;
			this.proxy = createContext$1(this.object, this.manager);
		}
	}
	get(prop) {
		return this.proxy[prop];
	}
	set(prop, value) {
		this.proxy[prop] = value;
	}
	toObject() {
		return this.proxy;
	}
	update(next) {
		Object.assign(this.proxy, next);
	}
	apply(mapFn) {
		this.update(mapFn(this.proxy));
	}
	createEffect(fn) {
		createEffect$1(fn, this);
	}
	setHost(host, eventType) {
		this.manager.setHost(host, eventType);
	}
	open(effect) {
		this.manager.push(effect);
		return this.proxy;
	}
	close() {
		this.manager.pop();
	}
};
function createContext$1(root, manager) {
	return new Proxy(root, {
		get: (subject, prop, receiver) => {
			const value = Reflect.get(subject, prop, receiver);
			if (manager.isRunning() && isObservable$1(value)) manager.subscribe(prop, subject);
			return value;
		},
		set: (subject, prop, newValue, receiver) => {
			const didSet = Reflect.set(subject, prop, newValue, receiver);
			if (didSet && isObservable$1(newValue)) manager.runEffects(prop, subject);
			return didSet;
		}
	});
}
function isObservable$1(value) {
	switch (typeof value) {
		case "object":
		case "number":
		case "string":
		case "symbol":
		case "boolean":
		case "undefined": return true;
		default: return false;
	}
}
var DirectEffect = class {
	constructor(fn, ...scope) {
		this.effectFn = () => fn(...scope);
	}
	execute() {
		this.effectFn();
	}
};
function createTemplate$1(fragment, render) {
	Object.assign(fragment, { render });
	return fragment;
}
function renderForEffects$1(original, effectors, ...scope) {
	const fragment = original.cloneNode(true);
	Array.from(effectors.entries()).forEach(([label, mutations]) => {
		const site = fragment.querySelector(`[data-${label}]`);
		if (site) mutations.forEach((fn) => fn(site, fragment, ...scope));
	});
	return fragment;
}
var TemplateParser$1 = class TemplateParser {
	static {
		this.parser = new DOMParser();
	}
	constructor(docType) {
		this.docType = "text/html";
		this.plugins = [];
		if (docType) this.docType = docType;
	}
	use(plugin) {
		this.plugins = this.plugins.concat(plugin);
	}
	parse(template, params) {
		const postProcess = {};
		const stringToParse = template.map((s, i) => {
			if (i >= params.length) return [s];
			const param = params[i];
			const place = this.classifyPlace(i, template);
			const mutation = this.tryReplacements(place, param);
			if (mutation) {
				const post = postProcess[place.nodeLabel];
				if (post) post.push(mutation);
				else postProcess[place.nodeLabel] = [mutation];
				switch (place.kind) {
					case "attr value": return [s, `"" data-${place.nodeLabel}`];
					case "tag content": return [s, `data-${place.nodeLabel}`];
					case "element content": return [s, `<ins data-${place.nodeLabel}></ins>`];
				}
			} else {
				console.error("No match for template parameter", place, param);
				throw `Failed to render template parameter ${i} around ${s}`;
			}
			return [s];
		}).flat().join("");
		const doc = TemplateParser.parser.parseFromString(stringToParse, this.docType);
		const collection = doc.head.childElementCount ? doc.head.children : doc.body.children;
		const fragment = new DocumentFragment();
		fragment.replaceChildren(...collection);
		const effectors = /* @__PURE__ */ new Map();
		for (const label in postProcess) {
			const site = fragment.querySelector(`[data-${label}]`);
			if (site) postProcess[label].forEach((m) => {
				const effector = m.apply(site, fragment);
				if (effector) {
					let list = effectors.get(label);
					if (list) list.push(effector);
					else effectors.set(label, [effector]);
				}
			});
		}
		return createTemplate$1(fragment, (...scope) => renderForEffects$1(fragment, effectors, ...scope));
	}
	static {
		this.OPEN_RE = /<([a-zA-z][$a-zA-Z0-9.-]*)\s+[^>]*$/;
	}
	static {
		this.IN_TAG_RE = /^(\s+|[^<>]*|"[^"]*")*$/;
	}
	static {
		this.ATTR_RE = /([$.]?[a-zA-Z][$a-zA-Z0-9.-]*)=\s*$/;
	}
	static {
		this.CLOSE_RE = /[/]?>[^<]*$/;
	}
	classifyPlace(i, template) {
		let tagOpen = null;
		for (let prev = i; prev >= 0; prev--) {
			if (template[prev].match(TemplateParser.CLOSE_RE)) break;
			tagOpen = template[prev].match(TemplateParser.OPEN_RE);
			if (tagOpen) break;
			if (!template[prev].match(TemplateParser.IN_TAG_RE)) break;
		}
		if (tagOpen) {
			const tagAttr = template[i].match(TemplateParser.ATTR_RE);
			if (tagAttr) return {
				kind: "attr value",
				nodeLabel: `node${i}`,
				tagName: tagOpen[1],
				attrName: tagAttr[1]
			};
			return {
				kind: "tag content",
				nodeLabel: `node${i}`,
				tagName: tagOpen[1]
			};
		}
		return {
			kind: "element content",
			nodeLabel: `node${i}`
		};
	}
	tryReplacements(place, param) {
		const replacements = this.plugins;
		for (let i = 0; i < replacements.length; i++) {
			const sub = replacements[i];
			if (place.kind === sub.place && checkType$1(param, sub)) return sub.mutator(place, param);
		}
	}
};
function checkType$1(param, sub) {
	if (typeof sub.types === "function") return sub.types(param, sub);
	return sub.types.includes(typeof param);
}
var Mutation$1 = class {
	constructor(place) {
		this.place = place;
	}
	apply(_site, _fragment) {
		throw "abstract method 'apply' called";
	}
};
var ElementContentMutation$1 = class extends Mutation$1 {
	constructor(place, content) {
		super(place);
		this.content = content;
	}
	apply(site, fragment) {
		(site.parentNode || fragment).replaceChild(this.content, site);
	}
};
var AttributeValueMutation$1 = class extends Mutation$1 {
	constructor(place, text) {
		super(place);
		this.text = text;
		this.name = place.attrName;
	}
	apply(site) {
		site.setAttribute(this.name, this.text);
	}
};
var ElementContentEffect$1 = class extends Mutation$1 {
	constructor(place, fn) {
		super(place);
		this.fn = fn;
	}
	apply(_site, _fragment) {
		const key = this.place.nodeLabel;
		return (site, fragment, ...scope) => {
			const start = new Comment(` <<< ${key} `);
			const end = new Comment(` >>> ${key} `);
			const placeholder = new DocumentFragment();
			placeholder.replaceChildren(start, end);
			(site.parentNode || fragment).replaceChild(placeholder, site);
			createEffect$1((...args) => {
				replaceElementContent$1(this.fn(...args), start, end);
			}, ...scope);
		};
	}
};
function replaceElementContent$1(value, start, end) {
	const parent = start.parentNode;
	if (!parent) throw new Error("No parent for placeholder");
	const valueToNode = (v) => {
		if (Array.isArray(v)) {
			const frag = new DocumentFragment();
			const nodes = v.map(valueToNode);
			frag.replaceChildren(...nodes);
			return frag;
		} else if (v instanceof Node) return v;
		else return new Text(v?.toString() || "");
	};
	const node = valueToNode(value);
	let p = start.nextSibling;
	while (p && p !== end) {
		const old = p;
		p = p.nextSibling;
		parent.removeChild(old);
	}
	if (node) parent.insertBefore(node, end);
}
var AttributeValueEffect$1 = class extends Mutation$1 {
	constructor(place, fn) {
		super(place);
		this.fn = fn;
		this.name = place.attrName;
	}
	apply(_site, _fragment) {
		return (site, _, ...scope) => createEffect$1((...args) => {
			replaceAttributeValue$1(this.fn(...args), site, this.name);
		}, ...scope);
	}
};
function replaceAttributeValue$1(value, site, attrName) {
	const special = attrName.match(/^([.$])(.+)$/);
	if (special) {
		const [_, pre, name] = special;
		switch (pre) {
			case ".":
				site[name] = value;
				break;
			case "$":
				if ("viewModel" in site && site.viewModel instanceof Context$1) site.viewModel.set(name, value);
				break;
		}
	} else switch (typeof value) {
		case "string":
			site.setAttribute(attrName, value);
			break;
		case "undefined":
		case "object":
		case "boolean":
			if (value) site.setAttribute(attrName, attrName);
			else site.removeAttribute(attrName);
			break;
		default: site.setAttribute(attrName, value?.toString());
	}
}
var TagReferenceEffect$1 = class extends Mutation$1 {
	constructor(place, fn) {
		super(place);
		this.fn = fn;
	}
	apply(_site, _fragment) {
		return (site, _, ...scope) => createEffect$1((...args) => {
			const value = this.fn(...args);
			if (typeof value === "function") value(site);
		}, ...scope);
	}
};
new TemplateParser$1().use([
	{
		place: "element content",
		types: [
			"string",
			"number",
			"bigint",
			"symbol",
			"boolean"
		],
		mutator: (place, value) => new ElementContentMutation$1(place, new Text(value?.toString() || ""))
	},
	{
		place: "attr value",
		types: [
			"string",
			"number",
			"bigint",
			"symbol"
		],
		mutator: (place, value) => new AttributeValueMutation$1(place, value?.toString() || "")
	},
	{
		place: "element content",
		types: (param) => param instanceof Node,
		mutator: (place, value) => new ElementContentMutation$1(place, value)
	},
	{
		place: "element content",
		types: (param) => Array.isArray(param),
		mutator: (place, value) => {
			const fragment = new DocumentFragment();
			const nodes = value.map((v) => v instanceof Node ? v : new Text(v?.toString() || ""));
			fragment.append(...nodes);
			return new ElementContentMutation$1(place, fragment);
		}
	},
	{
		place: "element content",
		types: ["function"],
		mutator: (place, param) => new ElementContentEffect$1(place, param)
	},
	{
		place: "attr value",
		types: ["function"],
		mutator: (place, param) => new AttributeValueEffect$1(place, param)
	},
	{
		place: "tag content",
		types: ["function"],
		mutator: (place, param) => new TagReferenceEffect$1(place, param)
	}
]);
var Provider = class Provider extends HTMLElement {
	static {
		this.DISCOVERY_EVENT = "un-provider:discover";
	}
	static {
		this.REGISTRY_EVENT = "un-provider:register";
	}
	static {
		this.CHANGE_EVENT = "un-provider:change";
	}
	static {
		document.addEventListener(Provider.DISCOVERY_EVENT, (event) => {
			const [contextLabel, respondFn] = event.detail;
			const provider = registeredProvider(contextLabel);
			if (provider) respondFn(provider);
		});
		document.addEventListener(Provider.REGISTRY_EVENT, (event) => {
			const [contextLabel, provider] = event.detail;
			registerProvider(contextLabel, provider);
		});
	}
	constructor(init, label) {
		super();
		this.contextLabel = label;
		this.context = new Context$1(init);
		this.context.setHost(this, Provider.CHANGE_EVENT);
		this.addEventListener(Provider.DISCOVERY_EVENT, (event) => {
			const [contextLabel, respondFn] = event.detail;
			if (contextLabel === this.contextLabel) {
				event.stopPropagation();
				respondFn(this);
			}
		});
		const registryEvent = new CustomEvent(Provider.REGISTRY_EVENT, {
			bubbles: true,
			composed: true,
			detail: [this.contextLabel, this]
		});
		this.dispatchEvent(registryEvent);
	}
	attach(observer) {
		this.addEventListener(Provider.CHANGE_EVENT, observer);
		return this.context.toObject();
	}
	detach(observer) {
		this.removeEventListener(Provider.CHANGE_EVENT, observer);
	}
};
function discover(observer, contextLabel) {
	return new Promise((resolve, reject) => {
		const discoveryEvent = new CustomEvent(Provider.DISCOVERY_EVENT, {
			bubbles: true,
			composed: true,
			detail: [contextLabel, (provider) => provider ? resolve(provider) : reject()]
		});
		if (observer.isConnected) observer.dispatchEvent(discoveryEvent);
		else document.dispatchEvent(discoveryEvent);
	});
}
var registry = {};
function registerProvider(label, provider) {
	registry[label] = provider;
}
function registeredProvider(label) {
	return registry[label];
}
var Observer = class {
	constructor(contextLabel) {
		this.contextLabel = contextLabel;
	}
	observe(from, fn) {
		return new Promise((resolve, reject) => {
			if (this.provider) resolve(this.attachObserver(fn));
			else discover(from, this.contextLabel).then((provider) => {
				this.provider = provider;
				resolve(this.attachObserver(fn));
			}).catch((err) => reject(err));
		});
	}
	attachObserver(fn) {
		const init = this.provider.attach((ev) => {
			const { property, value } = ev.detail;
			const effect = new DirectEffect(fn, {
				property,
				value
			});
			console.log("📡Signal received:", property, value);
			effect.execute();
		});
		console.log("👀 Initial observation:", init);
		return init;
	}
};
var Service = class {
	constructor(update, context, eventType = "service:message", autostart = true) {
		this._pending = [];
		this._context = context;
		this._update = update;
		this._eventType = eventType;
		this._running = autostart;
	}
	attach(host) {
		host.addEventListener(this._eventType, (ev) => {
			ev.stopPropagation();
			const message = ev.detail;
			this.consume(message);
		});
	}
	start() {
		if (!this._running) {
			console.log(`Starting ${this._eventType} service`);
			this._running = true;
			this._pending.forEach((msg) => this.process(msg));
		}
	}
	consume(message) {
		if (message.length !== 0) {
			const trueMsg = message;
			if (this._running) this.process(trueMsg);
			else {
				console.log(`📥 Queueing ${this._eventType} message`, message);
				this._pending.push(trueMsg);
			}
		}
	}
	process(message) {
		const reactions = message[2];
		let next;
		try {
			console.log(`📤 Processing ${this._eventType} message`, message);
			next = this._update(message, this._context.toObject());
			console.log("⏭️ next state: ", next);
		} catch (err) {
			console.log(`‼️ Error while processing ${message[0]} message:`, err);
			if (reactions && reactions.onFailure) {
				const error = err instanceof Error ? err : new Error(err);
				reactions.onFailure(error);
			}
		}
		if (!next) return;
		if (!Array.isArray(next)) this._context.update(next);
		else {
			const [now, ...later] = next;
			later.forEach((promise) => promise.then((command) => this.consume(command)).catch((err) => {
				console.log(`‼️ Error while processing later ${this._eventType} command:`, err.message);
				if (reactions && reactions.onFailure) reactions.onFailure(err);
			}));
			this._context.update(now);
		}
		if (reactions && reactions.onSuccess) reactions.onSuccess();
	}
};
function fromService(target, contextLabel) {
	return new FromService(target, contextLabel);
}
var FromService = class {
	constructor(client, contextLabel) {
		this.client = client;
		this.observer = new Observer(contextLabel);
	}
	start(fn) {
		return this.observer.observe(this.client, (s) => {
			fn(s.property, s.value);
		});
	}
};
//#endregion
//#region ../view/dist/view.js
var MappedSource = class {
	constructor(source, mapping) {
		this.initialized = false;
		this.origin = source;
		const entries = mapEntries(mapping).map(([t, s]) => [s, t]);
		this.inverse = Object.fromEntries(entries);
	}
	mapObservation(source) {
		const entries = Object.entries(source).map(([s, v]) => {
			const t = this.inverse[s];
			return t ? [t, v] : [];
		});
		return Object.fromEntries(entries.filter((pair) => pair.length > 0));
	}
	start(fn) {
		const sfn = (s, value) => {
			const t = this.inverse[s];
			fn(t, value);
			this.initialized = true;
		};
		return this.origin.start(sfn).then((obs) => this.initialized ? {} : this.mapObservation(obs));
	}
};
function mapEntries(mapping) {
	return Object.entries(mapping).map(([k, v]) => [k, v]);
}
var SignalEvent = class extends CustomEvent {
	constructor(eventType, signal) {
		super(eventType, {
			bubbles: true,
			cancelable: true,
			composed: true,
			detail: signal
		});
	}
};
function createEffect(fn, ...scope) {
	const effect = { execute() {
		console.log("▶️ Executing created effect:", scope, fn);
		fn(...scope.map((cx) => cx.open(effect)));
		scope.forEach((cx) => cx.close());
	} };
	effect.execute();
}
var Scheduler = class Scheduler {
	constructor() {
		this.signals = /* @__PURE__ */ new WeakMap();
		this.scheduled = /* @__PURE__ */ new WeakSet();
	}
	static {
		this.scheduler = new Scheduler();
	}
	subscribe(scope, key, effect) {
		let signals = this.signals.get(scope);
		if (!signals) {
			signals = /* @__PURE__ */ new Map();
			this.signals.set(scope, signals);
		}
		let signal = signals.get(key);
		if (!signal) {
			signal = /* @__PURE__ */ new Set();
			signals.set(key, signal);
		}
		signal.add(effect);
	}
	scheduleEffects(scope, key) {
		const signals = this.signals.get(scope);
		if (!signals) return;
		const signal = signals.get(key);
		if (signal) {
			for (const effect of signal) if (!this.scheduled.has(effect)) {
				this.scheduled.add(effect);
				setTimeout(() => {
					this.scheduled.delete(effect);
					effect.execute();
				});
			}
		}
	}
};
var EffectsManager = class {
	constructor() {
		this.running = [];
		this.eventType = "un-effect:change";
	}
	isRunning() {
		return this.running.length > 0;
	}
	push(effect) {
		this.running.push(effect);
	}
	pop() {
		this.running.pop();
	}
	current() {
		const len = this.running.length;
		return len ? this.running[len - 1] : void 0;
	}
	subscribe(key, scope) {
		const current = this.current();
		if (current) Scheduler.scheduler.subscribe(scope, key, current);
	}
	runEffects(key, scope) {
		console.log("⚙️ Running effects for signal", key, scope);
		Scheduler.scheduler.scheduleEffects(scope, key);
		if (this.host) {
			const evt = new SignalEvent(this.eventType, {
				property: key,
				value: scope[key]
			});
			this.host.dispatchEvent(evt);
		}
	}
	setHost(host, eventType) {
		this.host = host;
		if (eventType) this.eventType = eventType;
	}
};
var Context = class {
	static {
		this.CHANGE_EVENT_TYPE = "un-context:change";
	}
	constructor(init, adoptedContext) {
		if (adoptedContext) {
			this.manager = adoptedContext.manager;
			this.object = adoptedContext.object;
			this.proxy = adoptedContext.proxy;
			this.update(init);
		} else {
			this.manager = new EffectsManager();
			this.object = init;
			this.proxy = createContext(this.object, this.manager);
		}
	}
	get(prop) {
		return this.proxy[prop];
	}
	set(prop, value) {
		this.proxy[prop] = value;
	}
	toObject() {
		return this.proxy;
	}
	update(next) {
		Object.assign(this.proxy, next);
	}
	apply(mapFn) {
		this.update(mapFn(this.proxy));
	}
	createEffect(fn) {
		createEffect(fn, this);
	}
	setHost(host, eventType) {
		this.manager.setHost(host, eventType);
	}
	open(effect) {
		this.manager.push(effect);
		return this.proxy;
	}
	close() {
		this.manager.pop();
	}
};
function createContext(root, manager) {
	return new Proxy(root, {
		get: (subject, prop, receiver) => {
			const value = Reflect.get(subject, prop, receiver);
			if (manager.isRunning() && isObservable(value)) manager.subscribe(prop, subject);
			return value;
		},
		set: (subject, prop, newValue, receiver) => {
			const didSet = Reflect.set(subject, prop, newValue, receiver);
			if (didSet && isObservable(newValue)) manager.runEffects(prop, subject);
			return didSet;
		}
	});
}
function isObservable(value) {
	switch (typeof value) {
		case "object":
		case "number":
		case "string":
		case "symbol":
		case "boolean":
		case "undefined": return true;
		default: return false;
	}
}
function createTemplate(fragment, render) {
	Object.assign(fragment, { render });
	return fragment;
}
function renderForEffects(original, effectors, ...scope) {
	const fragment = original.cloneNode(true);
	Array.from(effectors.entries()).forEach(([label, mutations]) => {
		const site = fragment.querySelector(`[data-${label}]`);
		if (site) mutations.forEach((fn) => fn(site, fragment, ...scope));
	});
	return fragment;
}
var TemplateParser = class TemplateParser {
	static {
		this.parser = new DOMParser();
	}
	constructor(docType) {
		this.docType = "text/html";
		this.plugins = [];
		if (docType) this.docType = docType;
	}
	use(plugin) {
		this.plugins = this.plugins.concat(plugin);
	}
	parse(template, params) {
		const postProcess = {};
		const stringToParse = template.map((s, i) => {
			if (i >= params.length) return [s];
			const param = params[i];
			const place = this.classifyPlace(i, template);
			const mutation = this.tryReplacements(place, param);
			if (mutation) {
				const post = postProcess[place.nodeLabel];
				if (post) post.push(mutation);
				else postProcess[place.nodeLabel] = [mutation];
				switch (place.kind) {
					case "attr value": return [s, `"" data-${place.nodeLabel}`];
					case "tag content": return [s, `data-${place.nodeLabel}`];
					case "element content": return [s, `<ins data-${place.nodeLabel}></ins>`];
				}
			} else {
				console.error("No match for template parameter", place, param);
				throw `Failed to render template parameter ${i} around ${s}`;
			}
			return [s];
		}).flat().join("");
		const doc = TemplateParser.parser.parseFromString(stringToParse, this.docType);
		const collection = doc.head.childElementCount ? doc.head.children : doc.body.children;
		const fragment = new DocumentFragment();
		fragment.replaceChildren(...collection);
		const effectors = /* @__PURE__ */ new Map();
		for (const label in postProcess) {
			const site = fragment.querySelector(`[data-${label}]`);
			if (site) postProcess[label].forEach((m) => {
				const effector = m.apply(site, fragment);
				if (effector) {
					let list = effectors.get(label);
					if (list) list.push(effector);
					else effectors.set(label, [effector]);
				}
			});
		}
		return createTemplate(fragment, (...scope) => renderForEffects(fragment, effectors, ...scope));
	}
	static {
		this.OPEN_RE = /<([a-zA-z][$a-zA-Z0-9.-]*)\s+[^>]*$/;
	}
	static {
		this.IN_TAG_RE = /^(\s+|[^<>]*|"[^"]*")*$/;
	}
	static {
		this.ATTR_RE = /([$.]?[a-zA-Z][$a-zA-Z0-9.-]*)=\s*$/;
	}
	static {
		this.CLOSE_RE = /[/]?>[^<]*$/;
	}
	classifyPlace(i, template) {
		let tagOpen = null;
		for (let prev = i; prev >= 0; prev--) {
			if (template[prev].match(TemplateParser.CLOSE_RE)) break;
			tagOpen = template[prev].match(TemplateParser.OPEN_RE);
			if (tagOpen) break;
			if (!template[prev].match(TemplateParser.IN_TAG_RE)) break;
		}
		if (tagOpen) {
			const tagAttr = template[i].match(TemplateParser.ATTR_RE);
			if (tagAttr) return {
				kind: "attr value",
				nodeLabel: `node${i}`,
				tagName: tagOpen[1],
				attrName: tagAttr[1]
			};
			return {
				kind: "tag content",
				nodeLabel: `node${i}`,
				tagName: tagOpen[1]
			};
		}
		return {
			kind: "element content",
			nodeLabel: `node${i}`
		};
	}
	tryReplacements(place, param) {
		const replacements = this.plugins;
		for (let i = 0; i < replacements.length; i++) {
			const sub = replacements[i];
			if (place.kind === sub.place && checkType(param, sub)) return sub.mutator(place, param);
		}
	}
};
function checkType(param, sub) {
	if (typeof sub.types === "function") return sub.types(param, sub);
	return sub.types.includes(typeof param);
}
var Mutation = class {
	constructor(place) {
		this.place = place;
	}
	apply(_site, _fragment) {
		throw "abstract method 'apply' called";
	}
};
var ElementContentMutation = class extends Mutation {
	constructor(place, content) {
		super(place);
		this.content = content;
	}
	apply(site, fragment) {
		(site.parentNode || fragment).replaceChild(this.content, site);
	}
};
var AttributeValueMutation = class extends Mutation {
	constructor(place, text) {
		super(place);
		this.text = text;
		this.name = place.attrName;
	}
	apply(site) {
		site.setAttribute(this.name, this.text);
	}
};
var ElementContentEffect = class extends Mutation {
	constructor(place, fn) {
		super(place);
		this.fn = fn;
	}
	apply(_site, _fragment) {
		const key = this.place.nodeLabel;
		return (site, fragment, ...scope) => {
			const start = new Comment(` <<< ${key} `);
			const end = new Comment(` >>> ${key} `);
			const placeholder = new DocumentFragment();
			placeholder.replaceChildren(start, end);
			(site.parentNode || fragment).replaceChild(placeholder, site);
			createEffect((...args) => {
				replaceElementContent(this.fn(...args), start, end);
			}, ...scope);
		};
	}
};
function replaceElementContent(value, start, end) {
	const parent = start.parentNode;
	if (!parent) throw new Error("No parent for placeholder");
	const valueToNode = (v) => {
		if (Array.isArray(v)) {
			const frag = new DocumentFragment();
			const nodes = v.map(valueToNode);
			frag.replaceChildren(...nodes);
			return frag;
		} else if (v instanceof Node) return v;
		else return new Text(v?.toString() || "");
	};
	const node = valueToNode(value);
	let p = start.nextSibling;
	while (p && p !== end) {
		const old = p;
		p = p.nextSibling;
		parent.removeChild(old);
	}
	if (node) parent.insertBefore(node, end);
}
var AttributeValueEffect = class extends Mutation {
	constructor(place, fn) {
		super(place);
		this.fn = fn;
		this.name = place.attrName;
	}
	apply(_site, _fragment) {
		return (site, _, ...scope) => createEffect((...args) => {
			replaceAttributeValue(this.fn(...args), site, this.name);
		}, ...scope);
	}
};
function replaceAttributeValue(value, site, attrName) {
	const special = attrName.match(/^([.$])(.+)$/);
	if (special) {
		const [_, pre, name] = special;
		switch (pre) {
			case ".":
				site[name] = value;
				break;
			case "$":
				if ("viewModel" in site && site.viewModel instanceof Context) site.viewModel.set(name, value);
				break;
		}
	} else switch (typeof value) {
		case "string":
			site.setAttribute(attrName, value);
			break;
		case "undefined":
		case "object":
		case "boolean":
			if (value) site.setAttribute(attrName, attrName);
			else site.removeAttribute(attrName);
			break;
		default: site.setAttribute(attrName, value?.toString());
	}
}
var TagReferenceEffect = class extends Mutation {
	constructor(place, fn) {
		super(place);
		this.fn = fn;
	}
	apply(_site, _fragment) {
		return (site, _, ...scope) => createEffect((...args) => {
			const value = this.fn(...args);
			if (typeof value === "function") value(site);
		}, ...scope);
	}
};
new TemplateParser().use([
	{
		place: "element content",
		types: [
			"string",
			"number",
			"bigint",
			"symbol",
			"boolean"
		],
		mutator: (place, value) => new ElementContentMutation(place, new Text(value?.toString() || ""))
	},
	{
		place: "attr value",
		types: [
			"string",
			"number",
			"bigint",
			"symbol"
		],
		mutator: (place, value) => new AttributeValueMutation(place, value?.toString() || "")
	},
	{
		place: "element content",
		types: (param) => param instanceof Node,
		mutator: (place, value) => new ElementContentMutation(place, value)
	},
	{
		place: "element content",
		types: (param) => Array.isArray(param),
		mutator: (place, value) => {
			const fragment = new DocumentFragment();
			const nodes = value.map((v) => v instanceof Node ? v : new Text(v?.toString() || ""));
			fragment.append(...nodes);
			return new ElementContentMutation(place, fragment);
		}
	},
	{
		place: "element content",
		types: ["function"],
		mutator: (place, param) => new ElementContentEffect(place, param)
	},
	{
		place: "attr value",
		types: ["function"],
		mutator: (place, param) => new AttributeValueEffect(place, param)
	},
	{
		place: "tag content",
		types: ["function"],
		mutator: (place, param) => new TagReferenceEffect(place, param)
	}
]);
var ViewModel = class extends Context {
	constructor(init, adoptedContext) {
		super(init, adoptedContext);
	}
	get $() {
		return this.toObject();
	}
	with(source, ...keys) {
		const mapping = Object.fromEntries(keys.map((k) => [k, k]));
		return this.merge(new MappedSource(source, mapping));
	}
	withCalculated(source, mapping) {
		return this.merge(new MappedSource(source, mapping));
	}
	withRenamed(source, renaming) {
		return this.merge(new MappedSource(source, renaming));
	}
	merge(source) {
		if (source) {
			const initialized = /* @__PURE__ */ new Map();
			source.start((name, value) => {
				this.set(name, value);
				initialized.set(name, true);
			}).then((firstObservation) => {
				Object.keys(firstObservation).filter((k) => !initialized.get(k)).forEach((k) => {
					this.set(k, firstObservation[k]);
				});
			});
		}
		return this;
	}
	render(template) {
		return template.render(this);
	}
};
function createViewModel(init) {
	if (init === void 0) return new ViewModel({});
	else return new ViewModel(init);
}
//#endregion
//#region src/store.ts
var store_exports = /* @__PURE__ */ __exportAll$2({
	CONTEXT_DEFAULT: () => STORE_CONTEXT_DEFAULT,
	Provider: () => StoreProvider,
	Service: () => StoreService,
	dispatch: () => dispatch
});
var STORE_CONTEXT_DEFAULT = "context:store";
var StoreService = class StoreService extends Service {
	static {
		this.EVENT_TYPE = "store:message";
	}
	constructor(context, update) {
		super((message, model) => update(model, message), context, StoreService.EVENT_TYPE);
	}
};
var StoreProvider = class extends Provider {
	constructor(updateFn, init) {
		super(init, STORE_CONTEXT_DEFAULT);
		this.viewModel = createViewModel({ authenticated: false }).with(fromAuth(this), "authenticated", "username", "token");
		this._updateFn = updateFn;
	}
	connectedCallback() {
		new StoreService(this.context, (model, message) => this._updateFn(model, message, this.viewModel.toObject())).attach(this);
	}
};
function dispatch(target, message) {
	console.log("📨 Dispatching message:", message, target);
	target.dispatchEvent(new message_exports.Dispatch(message, StoreService.EVENT_TYPE));
}
//#endregion
//#region src/fromStore.ts
function fromStore(target, contextLabel = STORE_CONTEXT_DEFAULT) {
	return fromService(target, contextLabel);
}
//#endregion
export { store_exports as Store, fromStore };
